{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Period 4 -  Analyses of ChIP Seq data\n",
    "\n",
    "\n",
    "# Abstract\n",
    "\n",
    "ChIP-seq is a protocol for inferring the locations of proteins bound or associated with DNA. The raw data looks quite different than DNA- or RNA-seq, in that the NGS reads form tall \"peaks\" at the locations where the proteins were tightly bound to DNA in the cells which were used to create the sample. More specifically, ChIP-seq results in two peaks of reads of different strands (plus/minus also referred to as Watson/Crick), as shown in [Figure 1](http://www.ncbi.nlm.nih.gov/pmc/articles/PMC2592715/figure/F1/) of the MACS manuscript: [Zhang 2008](#foot).\n",
    "\n",
    "Most materials come from the [PH525x](https://github.com/genomicsclass/labs/blob/master/chipseq/ChIPseq.Rmd) lab.\n",
    "\n",
    "#  Approaches to peak calling and merging\n",
    "\n",
    "## Peak calling\n",
    "\n",
    "In the first lab, we use the MACS software to call peaks. The code for this is in the [MACS.txt](https://github.com/genomicsclass/labs/blob/master/course7/MACS.txt) file.\n",
    "\n",
    "There are many different algorithms for calling peaks, which have varying performance on different kinds of experiments. As mentioned in the lecture, for ChIP of proteins with broad peaks (such as modified histones), algorithms other than those for detecting sharp peaks might perform better.\n",
    "\n",
    "## After peak calling\n",
    "\n",
    "A number of analyses might be of interest following peak calling. In this lab we will focus on differential binding across samples, by focusing on the peak regions and counting the number of ChIP-seq reads which fall into the peaks for each sample.\n",
    "\n",
    "Motif-finding is common ChIP-seq analysis which is not explored in this course, as we do not cover the basics of analysis of sequences. Motif-finding refers to the task of looking for common strings of DNA letters contained within peaks. These are biologically meaningful, as a number of proteins which are bound to DNA have conformations which make certain strings of DNA letters more preferable for binding. For more references, see the [Footnotes](#foot).\n",
    "\n",
    "#  Calling differential binding across samples\n",
    "\n",
    "The following lab will go over the functionality of the `DiffBind` package, mostly using code from the vignette. This package is useful for manipulating ChIP-seq signal in R, for comparing signal across files and for performing tests of differential binding.\n",
    "\n",
    "## Reading peak files into R\n",
    "\n",
    "The `DiffBind` package provides some called peaks that we can analyze. From the `DiffBind` vignette:\n",
    "\n",
    "> The dataset for this example consists of ChIPs against the transcription factor ERa using five breast cancer cell lines (C.S. Ross-Innes et al, Nature 481(7381):389–393, 2012). Three of these cell lines are responsive to tamoxifen treatment, while two others are resistant to tamoxifen. There are at least two replicates for each of the cell lines, with one cell line having three replicates, for a total of eleven sequenced libraries.\n",
    "\n",
    "> Of the five cell lines, two are based on MCF7 cells: the regular tamoxifen responsive line,\n",
    "as well as MCF7 cells specially treated with tamoxifen until a tamoxifen resistant cell line\n",
    "is obtained. For each sample, we have one peakset originally derived using the MACS peak\n",
    "caller (Y. Zhang et al, Genome Biol, 9(9):R137, 2008), for a total of eleven peaksets. Note that to save space in the package, only data for chromosome 18 is used for the vignette."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We're going to take this opportunity to demonstrate the use of BiocFileCache to organize and keep track of files:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "suppressPackageStartupMessages({\n",
    "    library(DiffBind)\n",
    "    library(BiocFileCache)\n",
    "})\n",
    "bfc <- BiocFileCache(cache=\"~/chipseq-data\")\n",
    "DBfiles <- list.files(system.file(\"extra\", package=\"DiffBind\"), \n",
    "                     recursive = TRUE, full.names = TRUE)\n",
    "DBfiles <- DBfiles[!DBfiles %in% bfcinfo(bfc)$rname]\n",
    "for (i in seq_along(DBfiles))\n",
    "  bfcadd(bfc, rname=DBfiles[i], rtype=\"local\", action=\"copy\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Peaks are represented as a `.bed` file per sample, in this example summarized in `tamoxifen.csv`:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table>\n",
       "<thead><tr><th scope=col>SampleID</th><th scope=col>Tissue</th><th scope=col>Factor</th><th scope=col>Condition</th><th scope=col>Treatment</th><th scope=col>Replicate</th><th scope=col>bamReads</th><th scope=col>ControlID</th><th scope=col>bamControl</th><th scope=col>Peaks</th><th scope=col>PeakCaller</th></tr></thead>\n",
       "<tbody>\n",
       "\t<tr><td>BT4741                     </td><td>BT474                      </td><td>ER                         </td><td>Resistant                  </td><td>Full-Media                 </td><td>1                          </td><td>reads/Chr18_BT474_ER_1.bam </td><td>BT474c                     </td><td>reads/Chr18_BT474_input.bam</td><td>peaks/BT474_ER_1.bed.gz    </td><td>bed                        </td></tr>\n",
       "\t<tr><td>BT4742                     </td><td>BT474                      </td><td>ER                         </td><td>Resistant                  </td><td>Full-Media                 </td><td>2                          </td><td>reads/Chr18_BT474_ER_2.bam </td><td>BT474c                     </td><td>reads/Chr18_BT474_input.bam</td><td>peaks/BT474_ER_2.bed.gz    </td><td>bed                        </td></tr>\n",
       "\t<tr><td>MCF71                      </td><td>MCF7                       </td><td>ER                         </td><td>Responsive                 </td><td>Full-Media                 </td><td>1                          </td><td>reads/Chr18_MCF7_ER_1.bam  </td><td>MCF7c                      </td><td>reads/Chr18_MCF7_input.bam </td><td>peaks/MCF7_ER_1.bed.gz     </td><td>bed                        </td></tr>\n",
       "\t<tr><td>MCF72                      </td><td>MCF7                       </td><td>ER                         </td><td>Responsive                 </td><td>Full-Media                 </td><td>2                          </td><td>reads/Chr18_MCF7_ER_2.bam  </td><td>MCF7c                      </td><td>reads/Chr18_MCF7_input.bam </td><td>peaks/MCF7_ER_2.bed.gz     </td><td>bed                        </td></tr>\n",
       "\t<tr><td>MCF73                      </td><td>MCF7                       </td><td>ER                         </td><td>Responsive                 </td><td>Full-Media                 </td><td>3                          </td><td>reads/Chr18_MCF7_ER_3.bam  </td><td>MCF7c                      </td><td>reads/Chr18_MCF7_input.bam </td><td>peaks/MCF7_ER_3.bed.gz     </td><td>bed                        </td></tr>\n",
       "\t<tr><td>T47D1                      </td><td>T47D                       </td><td>ER                         </td><td>Responsive                 </td><td>Full-Media                 </td><td>1                          </td><td>reads/Chr18_T47D_ER_1.bam  </td><td>T47Dc                      </td><td>reads/Chr18_T47D_input.bam </td><td>peaks/T47D_ER_1.bed.gz     </td><td>bed                        </td></tr>\n",
       "\t<tr><td>T47D2                      </td><td>T47D                       </td><td>ER                         </td><td>Responsive                 </td><td>Full-Media                 </td><td>2                          </td><td>reads/Chr18_T47D_ER_2.bam  </td><td>T47Dc                      </td><td>reads/Chr18_T47D_input.bam </td><td>peaks/T47D_ER_2.bed.gz     </td><td>bed                        </td></tr>\n",
       "\t<tr><td>MCF7r1                     </td><td>MCF7                       </td><td>ER                         </td><td>Resistant                  </td><td>Full-Media                 </td><td>1                          </td><td>reads/Chr18_TAMR_ER_1.bam  </td><td>TAMRc                      </td><td>reads/Chr18_TAMR_input.bam </td><td>peaks/TAMR_ER_1.bed.gz     </td><td>bed                        </td></tr>\n",
       "\t<tr><td>MCF7r2                     </td><td>MCF7                       </td><td>ER                         </td><td>Resistant                  </td><td>Full-Media                 </td><td>2                          </td><td>reads/Chr18_TAMR_ER_2.bam  </td><td>TAMRc                      </td><td>reads/Chr18_TAMR_input.bam </td><td>peaks/TAMR_ER_2.bed.gz     </td><td>bed                        </td></tr>\n",
       "\t<tr><td>ZR751                      </td><td>ZR75                       </td><td>ER                         </td><td>Responsive                 </td><td>Full-Media                 </td><td>1                          </td><td>reads/Chr18_ZR75_ER_1.bam  </td><td>ZR75c                      </td><td>reads/Chr18_ZR75_input.bam </td><td>peaks/ZR75_ER_1.bed.gz     </td><td>bed                        </td></tr>\n",
       "\t<tr><td>ZR752                      </td><td>ZR75                       </td><td>ER                         </td><td>Responsive                 </td><td>Full-Media                 </td><td>2                          </td><td>reads/Chr18_ZR75_ER_2.bam  </td><td>ZR75c                      </td><td>reads/Chr18_ZR75_input.bam </td><td>peaks/ZR75_ER_2.bed.gz     </td><td>bed                        </td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "\\begin{tabular}{r|lllllllllll}\n",
       " SampleID & Tissue & Factor & Condition & Treatment & Replicate & bamReads & ControlID & bamControl & Peaks & PeakCaller\\\\\n",
       "\\hline\n",
       "\t BT4741                           & BT474                            & ER                               & Resistant                        & Full-Media                       & 1                                & reads/Chr18\\_BT474\\_ER\\_1.bam & BT474c                           & reads/Chr18\\_BT474\\_input.bam  & peaks/BT474\\_ER\\_1.bed.gz      & bed                             \\\\\n",
       "\t BT4742                           & BT474                            & ER                               & Resistant                        & Full-Media                       & 2                                & reads/Chr18\\_BT474\\_ER\\_2.bam & BT474c                           & reads/Chr18\\_BT474\\_input.bam  & peaks/BT474\\_ER\\_2.bed.gz      & bed                             \\\\\n",
       "\t MCF71                            & MCF7                             & ER                               & Responsive                       & Full-Media                       & 1                                & reads/Chr18\\_MCF7\\_ER\\_1.bam  & MCF7c                            & reads/Chr18\\_MCF7\\_input.bam   & peaks/MCF7\\_ER\\_1.bed.gz       & bed                             \\\\\n",
       "\t MCF72                            & MCF7                             & ER                               & Responsive                       & Full-Media                       & 2                                & reads/Chr18\\_MCF7\\_ER\\_2.bam  & MCF7c                            & reads/Chr18\\_MCF7\\_input.bam   & peaks/MCF7\\_ER\\_2.bed.gz       & bed                             \\\\\n",
       "\t MCF73                            & MCF7                             & ER                               & Responsive                       & Full-Media                       & 3                                & reads/Chr18\\_MCF7\\_ER\\_3.bam  & MCF7c                            & reads/Chr18\\_MCF7\\_input.bam   & peaks/MCF7\\_ER\\_3.bed.gz       & bed                             \\\\\n",
       "\t T47D1                            & T47D                             & ER                               & Responsive                       & Full-Media                       & 1                                & reads/Chr18\\_T47D\\_ER\\_1.bam  & T47Dc                            & reads/Chr18\\_T47D\\_input.bam   & peaks/T47D\\_ER\\_1.bed.gz       & bed                             \\\\\n",
       "\t T47D2                            & T47D                             & ER                               & Responsive                       & Full-Media                       & 2                                & reads/Chr18\\_T47D\\_ER\\_2.bam  & T47Dc                            & reads/Chr18\\_T47D\\_input.bam   & peaks/T47D\\_ER\\_2.bed.gz       & bed                             \\\\\n",
       "\t MCF7r1                           & MCF7                             & ER                               & Resistant                        & Full-Media                       & 1                                & reads/Chr18\\_TAMR\\_ER\\_1.bam  & TAMRc                            & reads/Chr18\\_TAMR\\_input.bam   & peaks/TAMR\\_ER\\_1.bed.gz       & bed                             \\\\\n",
       "\t MCF7r2                           & MCF7                             & ER                               & Resistant                        & Full-Media                       & 2                                & reads/Chr18\\_TAMR\\_ER\\_2.bam  & TAMRc                            & reads/Chr18\\_TAMR\\_input.bam   & peaks/TAMR\\_ER\\_2.bed.gz       & bed                             \\\\\n",
       "\t ZR751                            & ZR75                             & ER                               & Responsive                       & Full-Media                       & 1                                & reads/Chr18\\_ZR75\\_ER\\_1.bam  & ZR75c                            & reads/Chr18\\_ZR75\\_input.bam   & peaks/ZR75\\_ER\\_1.bed.gz       & bed                             \\\\\n",
       "\t ZR752                            & ZR75                             & ER                               & Responsive                       & Full-Media                       & 2                                & reads/Chr18\\_ZR75\\_ER\\_2.bam  & ZR75c                            & reads/Chr18\\_ZR75\\_input.bam   & peaks/ZR75\\_ER\\_2.bed.gz       & bed                             \\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "SampleID | Tissue | Factor | Condition | Treatment | Replicate | bamReads | ControlID | bamControl | Peaks | PeakCaller | \n",
       "|---|---|---|---|---|---|---|---|---|---|---|\n",
       "| BT4741                      | BT474                       | ER                          | Resistant                   | Full-Media                  | 1                           | reads/Chr18_BT474_ER_1.bam  | BT474c                      | reads/Chr18_BT474_input.bam | peaks/BT474_ER_1.bed.gz     | bed                         | \n",
       "| BT4742                      | BT474                       | ER                          | Resistant                   | Full-Media                  | 2                           | reads/Chr18_BT474_ER_2.bam  | BT474c                      | reads/Chr18_BT474_input.bam | peaks/BT474_ER_2.bed.gz     | bed                         | \n",
       "| MCF71                       | MCF7                        | ER                          | Responsive                  | Full-Media                  | 1                           | reads/Chr18_MCF7_ER_1.bam   | MCF7c                       | reads/Chr18_MCF7_input.bam  | peaks/MCF7_ER_1.bed.gz      | bed                         | \n",
       "| MCF72                       | MCF7                        | ER                          | Responsive                  | Full-Media                  | 2                           | reads/Chr18_MCF7_ER_2.bam   | MCF7c                       | reads/Chr18_MCF7_input.bam  | peaks/MCF7_ER_2.bed.gz      | bed                         | \n",
       "| MCF73                       | MCF7                        | ER                          | Responsive                  | Full-Media                  | 3                           | reads/Chr18_MCF7_ER_3.bam   | MCF7c                       | reads/Chr18_MCF7_input.bam  | peaks/MCF7_ER_3.bed.gz      | bed                         | \n",
       "| T47D1                       | T47D                        | ER                          | Responsive                  | Full-Media                  | 1                           | reads/Chr18_T47D_ER_1.bam   | T47Dc                       | reads/Chr18_T47D_input.bam  | peaks/T47D_ER_1.bed.gz      | bed                         | \n",
       "| T47D2                       | T47D                        | ER                          | Responsive                  | Full-Media                  | 2                           | reads/Chr18_T47D_ER_2.bam   | T47Dc                       | reads/Chr18_T47D_input.bam  | peaks/T47D_ER_2.bed.gz      | bed                         | \n",
       "| MCF7r1                      | MCF7                        | ER                          | Resistant                   | Full-Media                  | 1                           | reads/Chr18_TAMR_ER_1.bam   | TAMRc                       | reads/Chr18_TAMR_input.bam  | peaks/TAMR_ER_1.bed.gz      | bed                         | \n",
       "| MCF7r2                      | MCF7                        | ER                          | Resistant                   | Full-Media                  | 2                           | reads/Chr18_TAMR_ER_2.bam   | TAMRc                       | reads/Chr18_TAMR_input.bam  | peaks/TAMR_ER_2.bed.gz      | bed                         | \n",
       "| ZR751                       | ZR75                        | ER                          | Responsive                  | Full-Media                  | 1                           | reads/Chr18_ZR75_ER_1.bam   | ZR75c                       | reads/Chr18_ZR75_input.bam  | peaks/ZR75_ER_1.bed.gz      | bed                         | \n",
       "| ZR752                       | ZR75                        | ER                          | Responsive                  | Full-Media                  | 2                           | reads/Chr18_ZR75_ER_2.bam   | ZR75c                       | reads/Chr18_ZR75_input.bam  | peaks/ZR75_ER_2.bed.gz      | bed                         | \n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "   SampleID Tissue Factor Condition  Treatment  Replicate\n",
       "1  BT4741   BT474  ER     Resistant  Full-Media 1        \n",
       "2  BT4742   BT474  ER     Resistant  Full-Media 2        \n",
       "3  MCF71    MCF7   ER     Responsive Full-Media 1        \n",
       "4  MCF72    MCF7   ER     Responsive Full-Media 2        \n",
       "5  MCF73    MCF7   ER     Responsive Full-Media 3        \n",
       "6  T47D1    T47D   ER     Responsive Full-Media 1        \n",
       "7  T47D2    T47D   ER     Responsive Full-Media 2        \n",
       "8  MCF7r1   MCF7   ER     Resistant  Full-Media 1        \n",
       "9  MCF7r2   MCF7   ER     Resistant  Full-Media 2        \n",
       "10 ZR751    ZR75   ER     Responsive Full-Media 1        \n",
       "11 ZR752    ZR75   ER     Responsive Full-Media 2        \n",
       "   bamReads                   ControlID bamControl                 \n",
       "1  reads/Chr18_BT474_ER_1.bam BT474c    reads/Chr18_BT474_input.bam\n",
       "2  reads/Chr18_BT474_ER_2.bam BT474c    reads/Chr18_BT474_input.bam\n",
       "3  reads/Chr18_MCF7_ER_1.bam  MCF7c     reads/Chr18_MCF7_input.bam \n",
       "4  reads/Chr18_MCF7_ER_2.bam  MCF7c     reads/Chr18_MCF7_input.bam \n",
       "5  reads/Chr18_MCF7_ER_3.bam  MCF7c     reads/Chr18_MCF7_input.bam \n",
       "6  reads/Chr18_T47D_ER_1.bam  T47Dc     reads/Chr18_T47D_input.bam \n",
       "7  reads/Chr18_T47D_ER_2.bam  T47Dc     reads/Chr18_T47D_input.bam \n",
       "8  reads/Chr18_TAMR_ER_1.bam  TAMRc     reads/Chr18_TAMR_input.bam \n",
       "9  reads/Chr18_TAMR_ER_2.bam  TAMRc     reads/Chr18_TAMR_input.bam \n",
       "10 reads/Chr18_ZR75_ER_1.bam  ZR75c     reads/Chr18_ZR75_input.bam \n",
       "11 reads/Chr18_ZR75_ER_2.bam  ZR75c     reads/Chr18_ZR75_input.bam \n",
       "   Peaks                   PeakCaller\n",
       "1  peaks/BT474_ER_1.bed.gz bed       \n",
       "2  peaks/BT474_ER_2.bed.gz bed       \n",
       "3  peaks/MCF7_ER_1.bed.gz  bed       \n",
       "4  peaks/MCF7_ER_2.bed.gz  bed       \n",
       "5  peaks/MCF7_ER_3.bed.gz  bed       \n",
       "6  peaks/T47D_ER_1.bed.gz  bed       \n",
       "7  peaks/T47D_ER_2.bed.gz  bed       \n",
       "8  peaks/TAMR_ER_1.bed.gz  bed       \n",
       "9  peaks/TAMR_ER_2.bed.gz  bed       \n",
       "10 peaks/ZR75_ER_1.bed.gz  bed       \n",
       "11 peaks/ZR75_ER_2.bed.gz  bed       "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "tamfile <- bfcquery(bfc, \"tamoxifen.csv\")$fpath\n",
    "read.csv(tamfile)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Just as a note, we now have the paths of all these files stored in our `BiocFileCache`:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<ol class=list-inline>\n",
       "\t<li>'/Library/Frameworks/R.framework/Versions/3.4/Resources/library/DiffBind/extra/peaks/BT474_ER_1.bed.gz'</li>\n",
       "\t<li>'/Library/Frameworks/R.framework/Versions/3.4/Resources/library/DiffBind/extra/peaks/BT474_ER_2.bed.gz'</li>\n",
       "\t<li>'/Library/Frameworks/R.framework/Versions/3.4/Resources/library/DiffBind/extra/peaks/MCF7_ER_1.bed.gz'</li>\n",
       "\t<li>'/Library/Frameworks/R.framework/Versions/3.4/Resources/library/DiffBind/extra/peaks/MCF7_ER_2.bed.gz'</li>\n",
       "\t<li>'/Library/Frameworks/R.framework/Versions/3.4/Resources/library/DiffBind/extra/peaks/MCF7_ER_3.bed.gz'</li>\n",
       "\t<li>'/Library/Frameworks/R.framework/Versions/3.4/Resources/library/DiffBind/extra/peaks/T47D_ER_1.bed.gz'</li>\n",
       "\t<li>'/Library/Frameworks/R.framework/Versions/3.4/Resources/library/DiffBind/extra/peaks/T47D_ER_2.bed.gz'</li>\n",
       "\t<li>'/Library/Frameworks/R.framework/Versions/3.4/Resources/library/DiffBind/extra/peaks/TAMR_ER_1.bed.gz'</li>\n",
       "\t<li>'/Library/Frameworks/R.framework/Versions/3.4/Resources/library/DiffBind/extra/peaks/TAMR_ER_2.bed.gz'</li>\n",
       "\t<li>'/Library/Frameworks/R.framework/Versions/3.4/Resources/library/DiffBind/extra/peaks/ZR75_ER_1.bed.gz'</li>\n",
       "\t<li>'/Library/Frameworks/R.framework/Versions/3.4/Resources/library/DiffBind/extra/peaks/ZR75_ER_2.bed.gz'</li>\n",
       "</ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item '/Library/Frameworks/R.framework/Versions/3.4/Resources/library/DiffBind/extra/peaks/BT474\\_ER\\_1.bed.gz'\n",
       "\\item '/Library/Frameworks/R.framework/Versions/3.4/Resources/library/DiffBind/extra/peaks/BT474\\_ER\\_2.bed.gz'\n",
       "\\item '/Library/Frameworks/R.framework/Versions/3.4/Resources/library/DiffBind/extra/peaks/MCF7\\_ER\\_1.bed.gz'\n",
       "\\item '/Library/Frameworks/R.framework/Versions/3.4/Resources/library/DiffBind/extra/peaks/MCF7\\_ER\\_2.bed.gz'\n",
       "\\item '/Library/Frameworks/R.framework/Versions/3.4/Resources/library/DiffBind/extra/peaks/MCF7\\_ER\\_3.bed.gz'\n",
       "\\item '/Library/Frameworks/R.framework/Versions/3.4/Resources/library/DiffBind/extra/peaks/T47D\\_ER\\_1.bed.gz'\n",
       "\\item '/Library/Frameworks/R.framework/Versions/3.4/Resources/library/DiffBind/extra/peaks/T47D\\_ER\\_2.bed.gz'\n",
       "\\item '/Library/Frameworks/R.framework/Versions/3.4/Resources/library/DiffBind/extra/peaks/TAMR\\_ER\\_1.bed.gz'\n",
       "\\item '/Library/Frameworks/R.framework/Versions/3.4/Resources/library/DiffBind/extra/peaks/TAMR\\_ER\\_2.bed.gz'\n",
       "\\item '/Library/Frameworks/R.framework/Versions/3.4/Resources/library/DiffBind/extra/peaks/ZR75\\_ER\\_1.bed.gz'\n",
       "\\item '/Library/Frameworks/R.framework/Versions/3.4/Resources/library/DiffBind/extra/peaks/ZR75\\_ER\\_2.bed.gz'\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. '/Library/Frameworks/R.framework/Versions/3.4/Resources/library/DiffBind/extra/peaks/BT474_ER_1.bed.gz'\n",
       "2. '/Library/Frameworks/R.framework/Versions/3.4/Resources/library/DiffBind/extra/peaks/BT474_ER_2.bed.gz'\n",
       "3. '/Library/Frameworks/R.framework/Versions/3.4/Resources/library/DiffBind/extra/peaks/MCF7_ER_1.bed.gz'\n",
       "4. '/Library/Frameworks/R.framework/Versions/3.4/Resources/library/DiffBind/extra/peaks/MCF7_ER_2.bed.gz'\n",
       "5. '/Library/Frameworks/R.framework/Versions/3.4/Resources/library/DiffBind/extra/peaks/MCF7_ER_3.bed.gz'\n",
       "6. '/Library/Frameworks/R.framework/Versions/3.4/Resources/library/DiffBind/extra/peaks/T47D_ER_1.bed.gz'\n",
       "7. '/Library/Frameworks/R.framework/Versions/3.4/Resources/library/DiffBind/extra/peaks/T47D_ER_2.bed.gz'\n",
       "8. '/Library/Frameworks/R.framework/Versions/3.4/Resources/library/DiffBind/extra/peaks/TAMR_ER_1.bed.gz'\n",
       "9. '/Library/Frameworks/R.framework/Versions/3.4/Resources/library/DiffBind/extra/peaks/TAMR_ER_2.bed.gz'\n",
       "10. '/Library/Frameworks/R.framework/Versions/3.4/Resources/library/DiffBind/extra/peaks/ZR75_ER_1.bed.gz'\n",
       "11. '/Library/Frameworks/R.framework/Versions/3.4/Resources/library/DiffBind/extra/peaks/ZR75_ER_2.bed.gz'\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       " [1] \"/Library/Frameworks/R.framework/Versions/3.4/Resources/library/DiffBind/extra/peaks/BT474_ER_1.bed.gz\"\n",
       " [2] \"/Library/Frameworks/R.framework/Versions/3.4/Resources/library/DiffBind/extra/peaks/BT474_ER_2.bed.gz\"\n",
       " [3] \"/Library/Frameworks/R.framework/Versions/3.4/Resources/library/DiffBind/extra/peaks/MCF7_ER_1.bed.gz\" \n",
       " [4] \"/Library/Frameworks/R.framework/Versions/3.4/Resources/library/DiffBind/extra/peaks/MCF7_ER_2.bed.gz\" \n",
       " [5] \"/Library/Frameworks/R.framework/Versions/3.4/Resources/library/DiffBind/extra/peaks/MCF7_ER_3.bed.gz\" \n",
       " [6] \"/Library/Frameworks/R.framework/Versions/3.4/Resources/library/DiffBind/extra/peaks/T47D_ER_1.bed.gz\" \n",
       " [7] \"/Library/Frameworks/R.framework/Versions/3.4/Resources/library/DiffBind/extra/peaks/T47D_ER_2.bed.gz\" \n",
       " [8] \"/Library/Frameworks/R.framework/Versions/3.4/Resources/library/DiffBind/extra/peaks/TAMR_ER_1.bed.gz\" \n",
       " [9] \"/Library/Frameworks/R.framework/Versions/3.4/Resources/library/DiffBind/extra/peaks/TAMR_ER_2.bed.gz\" \n",
       "[10] \"/Library/Frameworks/R.framework/Versions/3.4/Resources/library/DiffBind/extra/peaks/ZR75_ER_1.bed.gz\" \n",
       "[11] \"/Library/Frameworks/R.framework/Versions/3.4/Resources/library/DiffBind/extra/peaks/ZR75_ER_2.bed.gz\" "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "bfcquery(bfc, \"peaks\")$rpath"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Peak merging and differential binding setup\n",
    "\n",
    "The `dba` function creates the basic object for an analysis of *Differential Binding Affinity*. The sample sheet specifies a data frame of file with certain required columns. Note that columns have restricted names, including *Tissue*, *Factor*, *Condition*, etc., which will be referred to later in analysis.\n",
    "\n",
    "This function will automatically create a correlation plot showing the overlap of the peaks for all the samples."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "BT4741 BT474 ER Resistant Full-Media 1 bed\n",
      "BT4742 BT474 ER Resistant Full-Media 2 bed\n",
      "MCF71 MCF7 ER Responsive Full-Media 1 bed\n",
      "MCF72 MCF7 ER Responsive Full-Media 2 bed\n",
      "MCF73 MCF7 ER Responsive Full-Media 3 bed\n",
      "T47D1 T47D ER Responsive Full-Media 1 bed\n",
      "T47D2 T47D ER Responsive Full-Media 2 bed\n",
      "MCF7r1 MCF7 ER Resistant Full-Media 1 bed\n",
      "MCF7r2 MCF7 ER Resistant Full-Media 2 bed\n",
      "ZR751 ZR75 ER Responsive Full-Media 1 bed\n",
      "ZR752 ZR75 ER Responsive Full-Media 2 bed\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "11 Samples, 2845 sites in matrix (3795 total):\n",
       "       ID Tissue Factor  Condition  Treatment Replicate Caller Intervals\n",
       "1  BT4741  BT474     ER  Resistant Full-Media         1    bed      1080\n",
       "2  BT4742  BT474     ER  Resistant Full-Media         2    bed      1122\n",
       "3   MCF71   MCF7     ER Responsive Full-Media         1    bed      1556\n",
       "4   MCF72   MCF7     ER Responsive Full-Media         2    bed      1046\n",
       "5   MCF73   MCF7     ER Responsive Full-Media         3    bed      1339\n",
       "6   T47D1   T47D     ER Responsive Full-Media         1    bed       527\n",
       "7   T47D2   T47D     ER Responsive Full-Media         2    bed       373\n",
       "8  MCF7r1   MCF7     ER  Resistant Full-Media         1    bed      1438\n",
       "9  MCF7r2   MCF7     ER  Resistant Full-Media         2    bed       930\n",
       "10  ZR751   ZR75     ER Responsive Full-Media         1    bed      2346\n",
       "11  ZR752   ZR75     ER Responsive Full-Media         2    bed      2345"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "setwd(system.file(\"extra\", package=\"DiffBind\")) #necessary because `tamfile` contains relative paths\n",
    "ta <- dba(sampleSheet=tamfile)\n",
    "ta"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "From the `DiffBind` vignette, we have:\n",
    "\n",
    "> This shows how many peaks are in each peakset, as well as (in the first line) \n",
    "> total number of unique peaks *after merging overlapping ones* (3557) and the \n",
    "> default binding matrix of 11 samples by the 2602 sites that *overlap in at \n",
    "> least two of the samples*.\"\n",
    "\n",
    "We can access the peaks for each file:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<ol class=list-inline>\n",
       "\t<li>'config'</li>\n",
       "\t<li>'chrmap'</li>\n",
       "\t<li>'peaks'</li>\n",
       "\t<li>'class'</li>\n",
       "\t<li>'masks'</li>\n",
       "\t<li>'samples'</li>\n",
       "\t<li>'called'</li>\n",
       "\t<li>'totalMerged'</li>\n",
       "\t<li>'merged'</li>\n",
       "\t<li>'binding'</li>\n",
       "\t<li>'attributes'</li>\n",
       "\t<li>'minOverlap'</li>\n",
       "</ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 'config'\n",
       "\\item 'chrmap'\n",
       "\\item 'peaks'\n",
       "\\item 'class'\n",
       "\\item 'masks'\n",
       "\\item 'samples'\n",
       "\\item 'called'\n",
       "\\item 'totalMerged'\n",
       "\\item 'merged'\n",
       "\\item 'binding'\n",
       "\\item 'attributes'\n",
       "\\item 'minOverlap'\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 'config'\n",
       "2. 'chrmap'\n",
       "3. 'peaks'\n",
       "4. 'class'\n",
       "5. 'masks'\n",
       "6. 'samples'\n",
       "7. 'called'\n",
       "8. 'totalMerged'\n",
       "9. 'merged'\n",
       "10. 'binding'\n",
       "11. 'attributes'\n",
       "12. 'minOverlap'\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       " [1] \"config\"      \"chrmap\"      \"peaks\"       \"class\"       \"masks\"      \n",
       " [6] \"samples\"     \"called\"      \"totalMerged\" \"merged\"      \"binding\"    \n",
       "[11] \"attributes\"  \"minOverlap\" "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "'list'"
      ],
      "text/latex": [
       "'list'"
      ],
      "text/markdown": [
       "'list'"
      ],
      "text/plain": [
       "[1] \"list\""
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<table>\n",
       "<thead><tr><th scope=col>V1</th><th scope=col>V2</th><th scope=col>V3</th><th scope=col>V5</th></tr></thead>\n",
       "<tbody>\n",
       "\t<tr><td>chr18     </td><td>215562    </td><td>216063    </td><td>0.01742633</td></tr>\n",
       "\t<tr><td>chr18     </td><td>311530    </td><td>312105    </td><td>0.06909971</td></tr>\n",
       "\t<tr><td>chr18     </td><td>356656    </td><td>357315    </td><td>0.02859148</td></tr>\n",
       "\t<tr><td>chr18     </td><td>371110    </td><td>372092    </td><td>0.03846775</td></tr>\n",
       "\t<tr><td>chr18     </td><td>395116    </td><td>396464    </td><td>0.47995863</td></tr>\n",
       "\t<tr><td>chr18     </td><td>399014    </td><td>400382    </td><td>0.56996310</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "\\begin{tabular}{r|llll}\n",
       " V1 & V2 & V3 & V5\\\\\n",
       "\\hline\n",
       "\t chr18      & 215562     & 216063     & 0.01742633\\\\\n",
       "\t chr18      & 311530     & 312105     & 0.06909971\\\\\n",
       "\t chr18      & 356656     & 357315     & 0.02859148\\\\\n",
       "\t chr18      & 371110     & 372092     & 0.03846775\\\\\n",
       "\t chr18      & 395116     & 396464     & 0.47995863\\\\\n",
       "\t chr18      & 399014     & 400382     & 0.56996310\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "V1 | V2 | V3 | V5 | \n",
       "|---|---|---|---|---|---|\n",
       "| chr18      | 215562     | 216063     | 0.01742633 | \n",
       "| chr18      | 311530     | 312105     | 0.06909971 | \n",
       "| chr18      | 356656     | 357315     | 0.02859148 | \n",
       "| chr18      | 371110     | 372092     | 0.03846775 | \n",
       "| chr18      | 395116     | 396464     | 0.47995863 | \n",
       "| chr18      | 399014     | 400382     | 0.56996310 | \n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "  V1    V2     V3     V5        \n",
       "1 chr18 215562 216063 0.01742633\n",
       "2 chr18 311530 312105 0.06909971\n",
       "3 chr18 356656 357315 0.02859148\n",
       "4 chr18 371110 372092 0.03846775\n",
       "5 chr18 395116 396464 0.47995863\n",
       "6 chr18 399014 400382 0.56996310"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "names(ta)\n",
    "class(ta$peaks)\n",
    "head(ta$peaks[[1]])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Differential binding\n",
    "\n",
    "The following code chunk will count the reads from the BAM files specified in the `samples` slot:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table>\n",
       "<thead><tr><th scope=col>SampleID</th><th scope=col>Tissue</th><th scope=col>Factor</th><th scope=col>Condition</th><th scope=col>Treatment</th><th scope=col>Replicate</th><th scope=col>bamReads</th><th scope=col>ControlID</th><th scope=col>bamControl</th><th scope=col>Peaks</th><th scope=col>PeakCaller</th></tr></thead>\n",
       "<tbody>\n",
       "\t<tr><td>BT4741                     </td><td>BT474                      </td><td>ER                         </td><td>Resistant                  </td><td>Full-Media                 </td><td>1                          </td><td>reads/Chr18_BT474_ER_1.bam </td><td>BT474c                     </td><td>reads/Chr18_BT474_input.bam</td><td>peaks/BT474_ER_1.bed.gz    </td><td>bed                        </td></tr>\n",
       "\t<tr><td>BT4742                     </td><td>BT474                      </td><td>ER                         </td><td>Resistant                  </td><td>Full-Media                 </td><td>2                          </td><td>reads/Chr18_BT474_ER_2.bam </td><td>BT474c                     </td><td>reads/Chr18_BT474_input.bam</td><td>peaks/BT474_ER_2.bed.gz    </td><td>bed                        </td></tr>\n",
       "\t<tr><td>MCF71                      </td><td>MCF7                       </td><td>ER                         </td><td>Responsive                 </td><td>Full-Media                 </td><td>1                          </td><td>reads/Chr18_MCF7_ER_1.bam  </td><td>MCF7c                      </td><td>reads/Chr18_MCF7_input.bam </td><td>peaks/MCF7_ER_1.bed.gz     </td><td>bed                        </td></tr>\n",
       "\t<tr><td>MCF72                      </td><td>MCF7                       </td><td>ER                         </td><td>Responsive                 </td><td>Full-Media                 </td><td>2                          </td><td>reads/Chr18_MCF7_ER_2.bam  </td><td>MCF7c                      </td><td>reads/Chr18_MCF7_input.bam </td><td>peaks/MCF7_ER_2.bed.gz     </td><td>bed                        </td></tr>\n",
       "\t<tr><td>MCF73                      </td><td>MCF7                       </td><td>ER                         </td><td>Responsive                 </td><td>Full-Media                 </td><td>3                          </td><td>reads/Chr18_MCF7_ER_3.bam  </td><td>MCF7c                      </td><td>reads/Chr18_MCF7_input.bam </td><td>peaks/MCF7_ER_3.bed.gz     </td><td>bed                        </td></tr>\n",
       "\t<tr><td>T47D1                      </td><td>T47D                       </td><td>ER                         </td><td>Responsive                 </td><td>Full-Media                 </td><td>1                          </td><td>reads/Chr18_T47D_ER_1.bam  </td><td>T47Dc                      </td><td>reads/Chr18_T47D_input.bam </td><td>peaks/T47D_ER_1.bed.gz     </td><td>bed                        </td></tr>\n",
       "\t<tr><td>T47D2                      </td><td>T47D                       </td><td>ER                         </td><td>Responsive                 </td><td>Full-Media                 </td><td>2                          </td><td>reads/Chr18_T47D_ER_2.bam  </td><td>T47Dc                      </td><td>reads/Chr18_T47D_input.bam </td><td>peaks/T47D_ER_2.bed.gz     </td><td>bed                        </td></tr>\n",
       "\t<tr><td>MCF7r1                     </td><td>MCF7                       </td><td>ER                         </td><td>Resistant                  </td><td>Full-Media                 </td><td>1                          </td><td>reads/Chr18_TAMR_ER_1.bam  </td><td>TAMRc                      </td><td>reads/Chr18_TAMR_input.bam </td><td>peaks/TAMR_ER_1.bed.gz     </td><td>bed                        </td></tr>\n",
       "\t<tr><td>MCF7r2                     </td><td>MCF7                       </td><td>ER                         </td><td>Resistant                  </td><td>Full-Media                 </td><td>2                          </td><td>reads/Chr18_TAMR_ER_2.bam  </td><td>TAMRc                      </td><td>reads/Chr18_TAMR_input.bam </td><td>peaks/TAMR_ER_2.bed.gz     </td><td>bed                        </td></tr>\n",
       "\t<tr><td>ZR751                      </td><td>ZR75                       </td><td>ER                         </td><td>Responsive                 </td><td>Full-Media                 </td><td>1                          </td><td>reads/Chr18_ZR75_ER_1.bam  </td><td>ZR75c                      </td><td>reads/Chr18_ZR75_input.bam </td><td>peaks/ZR75_ER_1.bed.gz     </td><td>bed                        </td></tr>\n",
       "\t<tr><td>ZR752                      </td><td>ZR75                       </td><td>ER                         </td><td>Responsive                 </td><td>Full-Media                 </td><td>2                          </td><td>reads/Chr18_ZR75_ER_2.bam  </td><td>ZR75c                      </td><td>reads/Chr18_ZR75_input.bam </td><td>peaks/ZR75_ER_2.bed.gz     </td><td>bed                        </td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "\\begin{tabular}{r|lllllllllll}\n",
       " SampleID & Tissue & Factor & Condition & Treatment & Replicate & bamReads & ControlID & bamControl & Peaks & PeakCaller\\\\\n",
       "\\hline\n",
       "\t BT4741                           & BT474                            & ER                               & Resistant                        & Full-Media                       & 1                                & reads/Chr18\\_BT474\\_ER\\_1.bam & BT474c                           & reads/Chr18\\_BT474\\_input.bam  & peaks/BT474\\_ER\\_1.bed.gz      & bed                             \\\\\n",
       "\t BT4742                           & BT474                            & ER                               & Resistant                        & Full-Media                       & 2                                & reads/Chr18\\_BT474\\_ER\\_2.bam & BT474c                           & reads/Chr18\\_BT474\\_input.bam  & peaks/BT474\\_ER\\_2.bed.gz      & bed                             \\\\\n",
       "\t MCF71                            & MCF7                             & ER                               & Responsive                       & Full-Media                       & 1                                & reads/Chr18\\_MCF7\\_ER\\_1.bam  & MCF7c                            & reads/Chr18\\_MCF7\\_input.bam   & peaks/MCF7\\_ER\\_1.bed.gz       & bed                             \\\\\n",
       "\t MCF72                            & MCF7                             & ER                               & Responsive                       & Full-Media                       & 2                                & reads/Chr18\\_MCF7\\_ER\\_2.bam  & MCF7c                            & reads/Chr18\\_MCF7\\_input.bam   & peaks/MCF7\\_ER\\_2.bed.gz       & bed                             \\\\\n",
       "\t MCF73                            & MCF7                             & ER                               & Responsive                       & Full-Media                       & 3                                & reads/Chr18\\_MCF7\\_ER\\_3.bam  & MCF7c                            & reads/Chr18\\_MCF7\\_input.bam   & peaks/MCF7\\_ER\\_3.bed.gz       & bed                             \\\\\n",
       "\t T47D1                            & T47D                             & ER                               & Responsive                       & Full-Media                       & 1                                & reads/Chr18\\_T47D\\_ER\\_1.bam  & T47Dc                            & reads/Chr18\\_T47D\\_input.bam   & peaks/T47D\\_ER\\_1.bed.gz       & bed                             \\\\\n",
       "\t T47D2                            & T47D                             & ER                               & Responsive                       & Full-Media                       & 2                                & reads/Chr18\\_T47D\\_ER\\_2.bam  & T47Dc                            & reads/Chr18\\_T47D\\_input.bam   & peaks/T47D\\_ER\\_2.bed.gz       & bed                             \\\\\n",
       "\t MCF7r1                           & MCF7                             & ER                               & Resistant                        & Full-Media                       & 1                                & reads/Chr18\\_TAMR\\_ER\\_1.bam  & TAMRc                            & reads/Chr18\\_TAMR\\_input.bam   & peaks/TAMR\\_ER\\_1.bed.gz       & bed                             \\\\\n",
       "\t MCF7r2                           & MCF7                             & ER                               & Resistant                        & Full-Media                       & 2                                & reads/Chr18\\_TAMR\\_ER\\_2.bam  & TAMRc                            & reads/Chr18\\_TAMR\\_input.bam   & peaks/TAMR\\_ER\\_2.bed.gz       & bed                             \\\\\n",
       "\t ZR751                            & ZR75                             & ER                               & Responsive                       & Full-Media                       & 1                                & reads/Chr18\\_ZR75\\_ER\\_1.bam  & ZR75c                            & reads/Chr18\\_ZR75\\_input.bam   & peaks/ZR75\\_ER\\_1.bed.gz       & bed                             \\\\\n",
       "\t ZR752                            & ZR75                             & ER                               & Responsive                       & Full-Media                       & 2                                & reads/Chr18\\_ZR75\\_ER\\_2.bam  & ZR75c                            & reads/Chr18\\_ZR75\\_input.bam   & peaks/ZR75\\_ER\\_2.bed.gz       & bed                             \\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "SampleID | Tissue | Factor | Condition | Treatment | Replicate | bamReads | ControlID | bamControl | Peaks | PeakCaller | \n",
       "|---|---|---|---|---|---|---|---|---|---|---|\n",
       "| BT4741                      | BT474                       | ER                          | Resistant                   | Full-Media                  | 1                           | reads/Chr18_BT474_ER_1.bam  | BT474c                      | reads/Chr18_BT474_input.bam | peaks/BT474_ER_1.bed.gz     | bed                         | \n",
       "| BT4742                      | BT474                       | ER                          | Resistant                   | Full-Media                  | 2                           | reads/Chr18_BT474_ER_2.bam  | BT474c                      | reads/Chr18_BT474_input.bam | peaks/BT474_ER_2.bed.gz     | bed                         | \n",
       "| MCF71                       | MCF7                        | ER                          | Responsive                  | Full-Media                  | 1                           | reads/Chr18_MCF7_ER_1.bam   | MCF7c                       | reads/Chr18_MCF7_input.bam  | peaks/MCF7_ER_1.bed.gz      | bed                         | \n",
       "| MCF72                       | MCF7                        | ER                          | Responsive                  | Full-Media                  | 2                           | reads/Chr18_MCF7_ER_2.bam   | MCF7c                       | reads/Chr18_MCF7_input.bam  | peaks/MCF7_ER_2.bed.gz      | bed                         | \n",
       "| MCF73                       | MCF7                        | ER                          | Responsive                  | Full-Media                  | 3                           | reads/Chr18_MCF7_ER_3.bam   | MCF7c                       | reads/Chr18_MCF7_input.bam  | peaks/MCF7_ER_3.bed.gz      | bed                         | \n",
       "| T47D1                       | T47D                        | ER                          | Responsive                  | Full-Media                  | 1                           | reads/Chr18_T47D_ER_1.bam   | T47Dc                       | reads/Chr18_T47D_input.bam  | peaks/T47D_ER_1.bed.gz      | bed                         | \n",
       "| T47D2                       | T47D                        | ER                          | Responsive                  | Full-Media                  | 2                           | reads/Chr18_T47D_ER_2.bam   | T47Dc                       | reads/Chr18_T47D_input.bam  | peaks/T47D_ER_2.bed.gz      | bed                         | \n",
       "| MCF7r1                      | MCF7                        | ER                          | Resistant                   | Full-Media                  | 1                           | reads/Chr18_TAMR_ER_1.bam   | TAMRc                       | reads/Chr18_TAMR_input.bam  | peaks/TAMR_ER_1.bed.gz      | bed                         | \n",
       "| MCF7r2                      | MCF7                        | ER                          | Resistant                   | Full-Media                  | 2                           | reads/Chr18_TAMR_ER_2.bam   | TAMRc                       | reads/Chr18_TAMR_input.bam  | peaks/TAMR_ER_2.bed.gz      | bed                         | \n",
       "| ZR751                       | ZR75                        | ER                          | Responsive                  | Full-Media                  | 1                           | reads/Chr18_ZR75_ER_1.bam   | ZR75c                       | reads/Chr18_ZR75_input.bam  | peaks/ZR75_ER_1.bed.gz      | bed                         | \n",
       "| ZR752                       | ZR75                        | ER                          | Responsive                  | Full-Media                  | 2                           | reads/Chr18_ZR75_ER_2.bam   | ZR75c                       | reads/Chr18_ZR75_input.bam  | peaks/ZR75_ER_2.bed.gz      | bed                         | \n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "   SampleID Tissue Factor Condition  Treatment  Replicate\n",
       "1  BT4741   BT474  ER     Resistant  Full-Media 1        \n",
       "2  BT4742   BT474  ER     Resistant  Full-Media 2        \n",
       "3  MCF71    MCF7   ER     Responsive Full-Media 1        \n",
       "4  MCF72    MCF7   ER     Responsive Full-Media 2        \n",
       "5  MCF73    MCF7   ER     Responsive Full-Media 3        \n",
       "6  T47D1    T47D   ER     Responsive Full-Media 1        \n",
       "7  T47D2    T47D   ER     Responsive Full-Media 2        \n",
       "8  MCF7r1   MCF7   ER     Resistant  Full-Media 1        \n",
       "9  MCF7r2   MCF7   ER     Resistant  Full-Media 2        \n",
       "10 ZR751    ZR75   ER     Responsive Full-Media 1        \n",
       "11 ZR752    ZR75   ER     Responsive Full-Media 2        \n",
       "   bamReads                   ControlID bamControl                 \n",
       "1  reads/Chr18_BT474_ER_1.bam BT474c    reads/Chr18_BT474_input.bam\n",
       "2  reads/Chr18_BT474_ER_2.bam BT474c    reads/Chr18_BT474_input.bam\n",
       "3  reads/Chr18_MCF7_ER_1.bam  MCF7c     reads/Chr18_MCF7_input.bam \n",
       "4  reads/Chr18_MCF7_ER_2.bam  MCF7c     reads/Chr18_MCF7_input.bam \n",
       "5  reads/Chr18_MCF7_ER_3.bam  MCF7c     reads/Chr18_MCF7_input.bam \n",
       "6  reads/Chr18_T47D_ER_1.bam  T47Dc     reads/Chr18_T47D_input.bam \n",
       "7  reads/Chr18_T47D_ER_2.bam  T47Dc     reads/Chr18_T47D_input.bam \n",
       "8  reads/Chr18_TAMR_ER_1.bam  TAMRc     reads/Chr18_TAMR_input.bam \n",
       "9  reads/Chr18_TAMR_ER_2.bam  TAMRc     reads/Chr18_TAMR_input.bam \n",
       "10 reads/Chr18_ZR75_ER_1.bam  ZR75c     reads/Chr18_ZR75_input.bam \n",
       "11 reads/Chr18_ZR75_ER_2.bam  ZR75c     reads/Chr18_ZR75_input.bam \n",
       "   Peaks                   PeakCaller\n",
       "1  peaks/BT474_ER_1.bed.gz bed       \n",
       "2  peaks/BT474_ER_2.bed.gz bed       \n",
       "3  peaks/MCF7_ER_1.bed.gz  bed       \n",
       "4  peaks/MCF7_ER_2.bed.gz  bed       \n",
       "5  peaks/MCF7_ER_3.bed.gz  bed       \n",
       "6  peaks/T47D_ER_1.bed.gz  bed       \n",
       "7  peaks/T47D_ER_2.bed.gz  bed       \n",
       "8  peaks/TAMR_ER_1.bed.gz  bed       \n",
       "9  peaks/TAMR_ER_2.bed.gz  bed       \n",
       "10 peaks/ZR75_ER_1.bed.gz  bed       \n",
       "11 peaks/ZR75_ER_2.bed.gz  bed       "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "ta$samples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAlgAAAJYCAIAAAAxBA+LAAAABmJLR0QA/wD/AP+gvaeTAAAg\nAElEQVR4nOzdeVwTR/8H8EkCISHIIYocAl4IIoKK+oiiYguoVKsiVrS2tSoqBdsfnlioitQW\njyqP4oEHghYPRMVKqUqLWBXvWm+RQ0WqyCGHHCEkm98f2ydNwxVhMUA+71dfvJLZ2dlvFss3\nszs7w5JKpQQAAEBdsVUdQEvJy8tbunRp3759dXR0+Hy+tbX1woULc3JylNzd2dm5V69ejESy\nYMECFov16NEj+cL8/Pz+/fuzWKyFCxfiuwgAgAq1z0R4/vx5Gxub1NTUTZs2vXz58sWLF1u3\nbr1w4YK9vf3ly5dVHR3Jzc0dOXLk7du3V6xYsXXrVhaLpeqIAADUl4aqA2DekydPPvzwQ1tb\n25SUFD6fTxe6u7sPGjSof//+U6ZMycjIEAgEqgovKyvr/ffff/bs2ffffx8YGKiqMAAAgNYO\nE+GaNWvKysq2bt0qy4K0jh07LliwICgoKDExcdq0aYSQtLS0kJCQK1eu1NTUODg4LF261NPT\ns842G6jp7OwsFovj4+M///zz27dv5+fnNxDb/fv33dzc8vLytm3b9sUXX8hvysjICA4O/u23\n3yorK/v16/fFF198+umnLBZr+vTphw8f/vPPPx0cHOiaFEWZmJjo6upmZGQ0+Sy1EiKRKCws\nzNrams1u7RcnKIq6d++enZ1d6w+VEEJRVHp6emBgIJfLVXUsAK2etH0Ri8V6enoWFhaN1vz5\n5581NTWXLVtWUFBQWlq6ZcsWDofz3Xff0VuHDx/es2dPJWsOHDiwX79+hBA9Pb3aB5o/fz4h\n5OHDhzdu3DA0NORwODExMQp17t27Z2BgMG3atOzs7PLy8qioKC6Xu2jRIqlUGh8fTwhZvXq1\nrPLvv/9OCAkKCnr709PqhIaGqvbff/sWGhqq6t8wQBvQ3nqEz58/Ly0tHTZsWMPVampqfHx8\nBg0atG7dOrpk4cKF9+7dW7ly5YwZMywtLd+q5q1btwYPHnzr1i06HdbpwoULS5YsKSsr69ev\n34wZMxS2zps3z8TE5Mcff9TQ0CCEfP755/fv39+8ebOfn9+4ceO0tbUTEhJWrVpFV05ISCCE\n0J3ats7KyooQEhAQ4OTkpOpYGtG2eoSXL1/evHkzfXoBoBGqzsQMu3fvHiHE29u74WrJycmE\nkIiICPnCpKQkQsj27dulcj1CZWqyWKycnJz6jkX3CNlsNp/PHzJkCCFkxYoV8hWysrIIIYGB\ngbUj3Lt3r1QqnTJlCiHk6dOn9KYePXrY2Ng0dibahri4OEJIXFycqgNpb3BiAZTXBr7bvhUD\nAwNCSFVVVcPVHj58SAjp3bu3fKGpqSkhJDc3921rdu3a1dzcvOEj6ujonDlzJjEx0djYeN26\ndb/99pts04MHDwghYWFhLDlubm6EkIKCAkKIl5cX+V9H8M6dO9nZ2VOnTm34cAAAoKT2lgiN\njY0NDAwUHtqT+emnn1gs1nfffVdTU0MI0dLSkt9aUVFRu1CZmsqMR4iPjx8xYkTnzp337dsn\nlUo/+eQTOsnJDvHDDz/U/p6yfPlyQsj48eN5PB6dCOmf3t7ejZ8LAABQQntLhGw228PDIz09\nne5mKaCzyLhx4ywsLAghT58+ld+anp5OCLG1tZUvVL5mw2RdxrFjxy5cuPDly5ezZs2SSqWE\nkG7duhFCMjMz69tXR0fH3d39woULr1+/PnHihJ2d3VsdGgAAGtDeEiEhJDg4WEtLKyAgQCKR\nyJffvXs3NjZ28uTJAwYMcHV1FQgER44cka9w9OhRPT29sWPHyhcqX1N569ats7OzS0pKCg8P\nJ4Q4ODh07979xIkTYrFYVoe+Unr37l36rZeXl0QiiYiI+PPPP2uPtQEAgKZ7x/ck341jx47x\neDxXV9cLFy5UVFSUlpYePnzY2Nh42LBhxcXFdJ1t27YRQtauXVtcXFxaWrphwwYWi0Vft5T+\n+/EJ5WvWSfb4hHzhnTt3eDwel8u9efOmVCo9e/aspqbm559/npeXV1RUtH37dh6PFxAQIKtf\nUlLC5XK1tLRYLNaTJ0+YOEmtAsZ0tBCcWADltcMeISHE09Pzzz//7N69+8yZM/X19c3Nzbds\n2fLNN9+cO3dOX1+frvPFF1/ExcUlJCSYmJiYmZmdPHkyISFh1qxZtVtTvqby+vXrFxYWJhKJ\nvL29y8vL3dzcLly4kJub27NnTwsLi7179/73v//94YcfZPX19PTef//96urq0aNH05dSAQCA\nESwpZnxuI6KioubMmXPkyJGPPvpI1bEw5ujRox999FFcXBzGwTILJxZAee2zR9gu5eXlmZmZ\n1TcJHAAANE17m1mmXSotLS0qKvrvf/8bFBRETz0DAABMwV/VNsDa2losFk+dOlVhnm4AAGg+\nJMI2IC8vT9UhAAC0W7hHCAAAag2JEAAA1BoSIQAAqDVmEuGVK1euXLnytpvepW+//ZbFYhUW\nFtbetGTJEhaL9euvv9beZGdnJxul2b9/fxsbGyUPVN+s3wAA0NowM1iGXlW19rP5Eomkvk0A\nAACtQXMT4erVq+t8TXv27BkhpEOHDs08Smvw559/qjoEAABoAc2cq1SZQ8ydO7dpjf/444/D\nhg3T1dXV1tbu2bPnggUL8vPz6U0ODg7vv//+3r17BwwYoKWlpa+vP2PGjKKiItm+hw4dcnBw\n0NLSMjQ0nD59uq+vLyGkoKCg9lEWL15MCElOTq69qW/fvhwOR3ZEa2tr+nVFRUVwcHCvXr20\ntLRMTEx8fHwKCwulUqmjo6PsU3/wwQd05UuXLrm7u+vq6vL5/KFDhx47dkz+EKdOnaI/grGx\n8dy5c5csWUIIqaqqkkqlw4cP/89//vP8+XNXV9fOnTsrc04cHR13797dr18/LS0tIyOjgICA\nGzdujB8/Xk9Pj8/nu7i4NO0X0XIwN3QLwYkFUF5zE+GJEydOnDhB/+k/UUtCQsKNGzea1nJ0\ndDQhJDQ0tLCwsKSkZMeOHWw2e+bMmfRWBwcHDodjbW195cqV8vLy+Ph4TU3NadOm0VsPHDhA\nCFmyZElRUdGTJ09k05IxlQinT59uZWV17969ioqK8+fPW1paOjk50ZtCQ0OJ3FoTP//8s6am\n5rJlywoKCkpLS7ds2cLhcL777jt6a1xcHJvN/vbbb0tKSjIzM8eNG0fHKUuEAwcO7NevHyFE\nT09PmXPCYrH69+9/8+bN169ff/7553Rr33zzTVFRUXp6eq9evZr2u2g5+HvdQnBiAZTHzDJM\nw4cPHz58OCNNyYwfP17hD3e/fv0GDhxIv3ZwcCCE3L9/X74+n8+XSqVVVVWdOnWSj0ckEvXs\n2bPhRFif2omwoqKCw+EsXLhQ1sLu3bsJIXfv3pX+OxGKRCJTU1NZjqTNmzdPQ0Pj6dOnQqHQ\nyMjoww8/lG2qqqoyNTWVT4QsFmvIkCG3bt0Si8VKnpPMzEz67fXr1wkhQ4cOlVUODAxs4ISr\nBP5etxCcWADlMTNq9OLFixcvXiSE1NTUlJaWltTShDZPnTqVkZEhe5uRkfHixQv5tXYtLS3l\nF2o3NzevqqoqKyv79ddfCwsLP/30U9kmTU3NDz/8sOHD1dcjrF1TIpFwOJyDBw8mJSXR8dDX\nfu3s7BRqnj9//sWLFx9//LF84aRJk8RicVJS0sWLF/Pz8ydOnCjbxOPxPvjgA4VG4uPj+/fv\nz+FwlDkn5ubmdMonhNALTg0aNEi2VSAQNHwSAADUEDOJUCQSff311+bm5lwuV19f36CWJrQp\nlUpjY2O9vb3t7Oz09PSGDh2qkFD19PTk37LZbDqShw8fEkIUFu0zNzdvQgx16tChw/bt26ur\nqz/44AMDA4MxY8b88MMPr169ql2TjqR3797yhXSfLzc39/Hjx4QQS0vLBuLs2rWrfEmj56T2\nqVY4SwAAoICZxydWrVoVFhZGCDE1NbW0tKRzUjNNnz79+PHjGzZsWLt2bffu3dlsdv/+/eUr\nsFisOncUi8Xkf3lRpqKiovkhycyZM8fLyyspKSklJSUlJeXs2bNhYWFXrlyR9cZoNTU1hBAt\nLa3akWhpadFbFVaTqK6uln/L5XLl3zb5nAAAQH2YSYQ//vgjIWTFihVr165l5G9xVlbWkSNH\nfH19v/rqK1mhSCRSSAx1ovuCT548kS+kH+RgkJ6e3vTp06dPn04IiYmJmTVr1qFDh4KDg+Xr\nWFhYEEKePn06cuRIWWF6ejohxNbWlk7YChNqZ2Zm1nfE5pwTAACoDzOXRgsKCsj/pmhhpEH6\nil+XLl1kJc+fP28gScgbM2YMn88/fPiwrKSqqur48eOMBEYI+fnnnxVmonFzcyOE8Hg88u8+\nmaurq0AgOHLkiPzuR48e1dPTGzt2rLOzM5vNTkxMlG0qLi4+ffp0fcdtzjkBAID6MJMIrays\nCCHyozaayc7OrkePHnv37r19+3ZFRcUvv/wyZswYNptdXl5eVVXV8L4dO3YMCwtLSUkJDQ0t\nKSnJz8+fNWuWwiXH5hg9enSvXr0WL1585cqVysrKJ0+e/N///R/9tCIhxNjYmPyvY6evr79+\n/fqkpKTvvvuupKSkrKxs48aNp0+fDg8P19HR6dq16xdffHH48OHt27eXlZU9evTIy8urvLy8\nJc4JAADUh5lEuGzZMkJITEwMI60RQrS0tJKSkmxtbZ2dnU1NTTdu3BgaGrpr165Xr14NGDCg\n0d39/PyWLl168OBBIyOjgQMHmpqarl27lqnYtLW1f/nll169ek2YMEFXV3fgwIGVlZW//fab\nmZkZIWTKlCnDhw+fPHny+PHjCSFffPFFXFxcQkKCiYmJmZnZyZMnExISZs2aRTf13//+d926\ndZs2bTI0NHRxcbG2tp4/fz6Hw1G4rcjIOQEAgDqxpEzMAioUCn/44Yc1a9asWLFi0qRJenp6\nCtdIFcZwtrSPPvro6NGjvXr1anNjJp89e1ZaWmpvb99AHRaLtWnTphEjRryzqFrO0aNHP/ro\no7i4uKlTp6o6lnYFJxZAecwMluHz+fSLkJCQkJCQ2hUYSbfKo9e7WBn6zUTPCRShKKmEklKU\nlPrfa4mYEkukErFULKEkEqlYIiuhxGKpuIYSSyixWCoRUzViSlJD1UikkhqqpoYS10jE9E8x\nJRZRYjElFklqxJSkWiIRU1S1WFxDUSKJpIaiaiQS+oVYLJFIKIlYIpFI6NdisYSSUKIK0dNV\n5/j2RrqTetdI/t5aEVVBeun8MbxaKpaSGgkRS0kNRcTUPz/1u5D3vU7fvt0+EiEAgMoxkwin\nTJnCSDtqha3FEfQ3rrj9itvLgN1Lj6oQVae9kJZUc8ZZUA3s9v4UMnv546SkdxYnAED7xkwi\njI+PZ6SdtyKRSJKSkoRCYe1NlZWVhBCKaiihtAYdPa3Z+lrl555JTlQTLptjpqP1sbVEn0uo\n+ocdcTiEELN/P3QBAABNxkwiVIlz5841PHHaw/sPJ3tNbKCCyrE47A6jLbVHmssujYolFBEz\nNvgWAAAaxUwipJ8ZaEBeC/RgRo8e/dNPP9XZI/T19S0qKurTtw/jBwUAgHaGmUSoME+YUCgs\nLi6mKIrL5Q4cOJCeQoVxHA5nwoQJdW5avHhxUVERIzO9AQBA+8ZMIszNzVUoqaioOHny5Fdf\nfdW5c+dDhw4xchQAAADGtdQ9QoFAMGPGDIFAMGnSpGXLlm3bto3xQ7SDwTIAAKByLTtYxtXV\nlRASFxfXEomwHQyWAQAAlWvZRPjy5UvC9BJIMhgsAwAAzcdMInz69Gntwry8vKVLlxJC5NeR\nZxAGywAAQPMxkwi7d+9e3yYOh/PNN98wchQF7/IeYQ2LQ0hNDYtDSIuMgAUAAFVhJhGOGTOm\ndiGXy7W0tPz0008HDx7MyFEUvLN7hEcMByTq9+1X8fyOwHxpdpxxxYvmtwkAAK0EM4mwgeVk\nW847u0eYr9lBzGI/0jaRsNjFmjqNzB0AAABtCpODZd68eZOYmHjv3r2qqqouXbo4OTmNGDGC\nqTXra8M9QgAAaD7GEmFSUtInn3zy+vVr+UI7O7vo6GhHR0emjgIAAMAsZhLh3bt3PT09q6ur\nJ0yYMHHiRF1d3ezs7IMHD965c+e99967fv167969GTmQvEYHy1yUsjqzNIZKRbUrXOJov2ax\n3EX/pO2XHK1zXAOupIYjFY+uyKnjcCw2IaSAq1/QacDggluEwqgZAID2gJlE+O2331ZXVy9Z\nsmTDhg2ywmXLlvn6+kZGRn777bf79+9n5EDyGh0ss+thxkUN7Us1dSTCpXzjPJbGrZoSTfL3\nUg+H+aYH+ab06/5Vr3Rr5TkRS4MQcrazYzmHryssti5+xMBnAAAAVWMmEZ47d44Q8uWXX8oX\nslisFStWREZGpqSkMHIUBY0OliF9+9bXa5MQIiVEQojmPyX/3MuU1H9fU0LYhBCKhbuPAADt\nBDOJsKSkhNS1GJORkREhJD8/n5GjKGh0sAzBYBkAAGgMM4nQzMzs6dOnjx8/7tu3r3z5o0eP\nCCEdO3Zk5CgKGr1HSDDpNgAANIaZRDhp0qTw8PBFixYdO3ZMR0eHLqyurl6+fDn539TbjGv0\nHiG5f58QL/mCME3dBI62plRaxuLIl+/jmybyjGRvv+rspCmVLC688ZCr94tOjyINbdkmIYdL\nCPnToE+f1w8aOHKxfrcb/WdThCVlsTSEb2x/DSXiSoU6+WMW1+gbGx4IIORfCZvyXC7tM5wQ\nNpGIyZ7lpEpIvthEjmwh50829GEBAKBJmEmEQUFBJ0+ePHv2rKWl5dSpUy0tLfPz848fP56T\nk6Ovr7969WpGjqJAmXuECuWX2LxnLA1S6w7gDU3dCrnUmM/hE0KyuPr3tTrlcvXka0oJixDy\nVGDWcGyluhblsifvBUY1fANNoWIirOrmKO7QieLyifjNvw7ReyjpZPH3G8u+pLKSmPYkNo5I\nhAAALYGZRNipU6eLFy/6+/snJCRERkbKyh0dHaOionr16lXfjjk5OXv27Pnll19evnxZVlZm\naGjYo0cPDw+POXPm6OvrN3xQ3CMEAIDmY+yBelNT0+PHj798+fLatWuFhYXa2tr29vZ9a/XJ\n5F27ds3Nza1Dhw7Ozs7Ozs46OjpCofD58+c//PDDxo0bz507Z2Njw1R4AAAAdWJ4PUITE5OJ\nE5Wd53rx4sVLly4NCgpSmIZNLBb7+Ph8+eWXZ8+ebWD3+gbLUBRFj2IlFFVYJfzu6nWpVCol\nUqlU+tegYUSPK6u580YGRyySSqkcRzPSSU+hnau5b/I66RJBHYeuFpPkTCKRsimphoRiU1JK\nIpVSUqmEosQUVZZXVq3Ll6//ssyQXUZJpZSEkkopiqKkUqmUIhqEkDekh6TkNUtfi2JTFEdK\nUZSUrfXPntpmhFNNCCECI9J1EJFIiX5X+hQ1cGYAAOAtSJunurp6x44dH3zwgXxh796958yZ\nc/ny5Yb35fP5lZWVdW7Ky8sTCAQN756cnNzIZwsNJVJpu/xvpI/PW/2aWq24uDhCSFxcnKoD\naW9wYgGU16weYXZ29oQJEx48UBw/+fjx48ePH+/du3f27Nnbt2/X0tKqc3d9ff1Hjx4NGDCg\n9qb8/HzZ6NP61DdYhqKoM2fOnC8qMjMxIT/+KL+psLDw4fXrffv3N+zShbBYUrmeKK+iQsTj\nsaRSis3WEIsJRdXweIWFhQ/v3LG1tTXr0EHE47ElEsJiaYhEQkFd/UTZIR4+7NOnj7GuLsXh\ncCQSiYYGVdfdSpZUSqTSgtev6fqdOnX6u5wQtljMpigJh0NxOIQQtkRCv/i7/ZSUzz09Gz45\nAACgrCan0NLS0p49exJC7O3tDx8+LL/p8ePHa9asoUe7TJ06tb4WgoKCzM3N9+7d+/TpU4lE\nIpVKJRLJ8+fPo6OjLSwsli5d2uTY6vO2X5Ob8LW6pQ/Rzr7pt7OP03rgxAIor+k9wq1bt2Zl\nZQ0aNCg1NVXw7x6SlZXVN998M3369BEjRhw9ejQhIWHSpEm1W1izZo1EIvHz8xMKhWw2W0tL\nSygUSqVSPp8/d+7cb7/9tsmxAQAAKKnpifDEiROEkJCQEEE91wl79er1/ffff/7553v37q0z\nEbLZ7O+//z4wMPDChQsZGRllZWUGBgZdu3Z1dXXV1dVtcmAAAADKa3oiTE9PJ4Q0vNbgBx98\nQAi5fv16A3X09PTGjx/f5DAAAACao+mPnItEIkJIw4+96+npEUIUVusFAABoPZqeCOmVJbKz\nsxuok5OTQxpLlgAAACrU9ETo7OxMCImNjW2gzvHjxwkhgwcPbvJRAAAAWlTTE6GPjw8hZOPG\njRcuXKizwh9//LF27VpCyOzZs5t8FGbx+XzZz5ao/w4O0YSQAACgIc159oLOcJqamgEBATdu\n3BCJRFKplKKo+/fvf/311zwejxAyfvx4Zh70YIJYLD5z5oxYLG6h+u/gEE0IqTXD424tBCcW\nQHnNmllm165dWlpaO3bs2Lx58+bNm1kslra2tlAolEgkdIWpU6fu37+/ubmaORwOx93dveXq\nv4NDNCEkAABoQLMWKuJwONu3b09LS5sxY4axsbFUKq2oqKAoyszM7OOPPz537lxcXBzdLwQA\nAGidGFh9wsnJycnJiRAiFAorKir09PQ0NBhe1AIAAKCFMJmxeDwe+n8AANC2YA13AABQa0iE\nAACg1pAIAQBArSERAgCAWkMiBAAAtYZECAAAag2JEAAA1BoSIQAAqDUkQgAAUGvtMxFSFBUc\nHGxmZiYQCEaOHHn9+vU6q1VVVS1ZsqRr167a2tq2trYbNmyQTReuwni6deumq6s7fPjw+ta3\nAgAABrXPRBgUFBQTE5OYmFhQUODm5ubq6pqTk1O7mr+/f3x8/MmTJwsKClasWBEYGPjdd9+p\nNp6kpKQzZ87k5OQMHTp0zJgx6enpLREPAAD8Q9XrQDGvpKSEz+fv3buXfkuvhvHNN9/Ursbl\ncvft2ycrmTx5soWFhariqa6u1tDQiIyMpN/W1NR06NBhzZo1jMfTqmDZvBaCEwugvHbYI7xw\n4UJVVdX7779Pv2WxWM7Ozr/88otCtczMTJFINGDAAFmJiYkJRVGqiqeiokIsFrNYLPqtVCqV\nSCQtEQ8AAMhrh4kwIyODzWabm5vLSkxMTF6+fKlQzdHRUSqVOjg40G/Ly8sTExPnzp2rqngM\nDAzGjh27YcOGO3fulJaWLlq0iKKoadOmMR4PAADIa4cLB5aVlWlpabHZ/+R4HR2dsrKyBnbJ\nzs6eOXPmtGnTgoODVRhPeHj48OHDZbl5/vz51tbWjMcDAADy2mEiNDAwEAqFFEXJco9QKDQ0\nNKyzslgsXr9+/ZkzZzZt2jR06FAVxvPixQtnZ+cRI0aEh4cbGhqePHly9uzZnTt3Dg0NbYmo\nmkwkEoWFhVlbW8un9iZLS0uT/WwmiqLS09MDAwO5XG7zWwMA9dEOE2HXrl2lUumLFy+6du1K\nl+Tm5vbo0aN2zYKCAk9PzzFjxqSmpspuzqkqnqNHjxYWFu7atatz586EkI8//vjcuXP79+9v\nbYlw/fr1q1atYrbN8PDw8PBwRppis9kt0a0HgHasHSZCV1dXPp//22+/ffbZZ4QQsVj8+++/\nL1q0qHbNTz75pGfPni39d1PJeOhMLD86hsvl6urqtmhsTWBlZUUICQgIcHJyan5rYrE4MTFx\n/PjxGhrN/ad4+fLlzZs30+G1Hsx2oJXHYFdbeeiUQ1ul2kGrLWThwoWWlpZ37tx59erVnDlz\njI2Ni4uLFercunWLEPLgwYNWEk9ubq6ent6ECROys7MrKip++ukngUCwY8eOdxDeW2m14/Jb\nZ2CtrUP/DoSGhqr6rAO8nXbYIySEbNy4UUtLy9XVtby8fNiwYefOndPX16c3eXl5Xbx4MS8v\n7/z584QQW1tb+R0NDQ0LCwtVEo+ZmdnVq1dDQkKGDx9eVlbWq1evrVu3fv7554wHA+8Ssx1o\n5THY1VZe6+yUAzRO1ZkY2pjW2fGSttbAWmdULUStPiy0J+3wOUIAAADlIRECAIBaa9uJMCUl\nZcaMGRYWFlpaWnp6eoMHD167dm3Dz863qFmzZrFYrHv37jW5BS8vLxaL9ejRIwajAgCABrTV\nwTISiWT+/Pl79+6VlYhEohs3bty4cSM6OvrMmTN1PjgIAACgoK32CBcuXLh3714tLa0VK1bc\nv39fKBS+fPny1KlTAwYMyMzMnDBhgkgkUnWMjUtISGCxWKtXr5aVxMfHS6VSGxsb1QUFAKBe\n2mQivHz58o4dOzQ1NU+fPv3dd9/Z2tpqaWkZGxuPHz/+999/79ev34MHD6KiolQdJgAAtAFt\nMhFGREQQQvz8/FxcXBQ26ejoBAUFEUISExNlha9fv168eHGPHj20tLS6dOni7e0tfxuPvrFX\nWFi4YcMGU1NTX1/f+goJIdXV1evWrbOzs+Pz+YaGhhMmTEhOTm4g1MLCwuXLl/fu3VtbW7tn\nz55z58599uwZvcnZ2Xny5MmEkJCQEBaLdeXKFVLXPcKGg6frC4XCH3/8sV+/fjwez8LC4uuv\nv24THWIAgNagTd4jTElJIYTMnDmzzq3Tpk2TX70oLy9vxIgRmZmZ9Nv8/PwjR46cPHny2LFj\nHh4esmohISF0fpVKpfUVVldXu7u7//777/RWoVCYmJiYmJi4bt26ZcuW1Y6koqJi6NChWVlZ\n9Nvs7Ozs7Oz4+PhLly717dtXmU+qZPA//PCDbKK458+ff//994SQ7777TplDAACoubbXIxQK\nhXl5eSwWy97eXpn6fn5+mZmZ7u7u9+7dq66ufvbs2eLFi4VC4SeffPLmzRtZtd27d2/atOnV\nq1c7d+6srzAkJOT333+fOHHiH3/8UVlZmZOTs3XrVoFAEBQUlJOTU/vQBw8ezMrKGjFixN27\nd6uqqp4+fTp//vzS0tKwsDBCyMWLF0+cOEEIWbVqlVQqrXPtCyWDX7NmzVlV+N0AACAASURB\nVPr161+9elVYWLhixQpCyIEDB5Q8nwAAaq7tJcKKigpCiLa2tqamZqOVX758mZCQ0LFjx2PH\njvXt25fL5VpYWGzcuHHChAmvX7+Wv3z6f//3fwEBAUZGRvK7yxeKxeLt27f37Nnz2LFjAwYM\n4PP55ubm/v7+S5YsEYvFdV4gvX37NiFk586ddnZ2PB7P0tLyhx9+IIQ8ffpUmU+qfPArVqxY\nunSpkZGRoaHh2rVrTU1Nc3Nza2pqlDkKAICaa3uJUE9Pj74rJn8Nsz7Xrl2jKGrq1Kk6Ojry\n5RMnTiSEyN9smzBhQu3d5Qvv379fWlqalZWloaHBkhMSEkIIycvLq717RESEVCq1tbWVSqWF\nhYV3796lh/AoE/lbBT9p0iTZaxaLZWlpSQiprKxU5igAAGqu7SVCDQ0NU1NTiUTy8OHDOitk\nZmayWCz6CQR6Bm06McijlzeqqqqSlejp6dVuSr6w4cm4q6uraxeKRKJNmzYNGTKEx+N17tzZ\n3t7+rdYiaHLw9Io/SqZbAAA11/YSISGEHiwaGxtb59ZDhw4RQpydnQkh2trahJD8/HyFOtnZ\n2YSQLl26KH9Qulv23nvv1Tln65o1a2rv8tlnny1evPj69esSicTCwqLRIaYKlA++5VYVBgBo\n99pkIpw7dy4hZNOmTfSagvKysrI2bdokq0MPqDlx4oT84wRSqfTgwYOEkLdaGadv374aGhrX\nrl0rKCiQL58+fTqLxbp06ZJC/ZKSksOHDxsZGaWkpFRWVj579uynn36S78Y1isHgAQCgPm0y\nEbq4uEydOlUoFLq4uGzYsCE7O1skEhUWFsbExIwYMaKkpOTzzz+nB2H27dvX0dHx2bNnEydO\nvH37tlAozMnJ+eKLL+7cudO3b98RI0Yof1AdHZ1p06aVl5dPmjTpxo0blZWVmZmZCxYsOHz4\nsL29/bBhwxTq09mLy+XyeDyxWJyXl7dv3z4vLy82m11eXk5fSqV7cs+ePZNIJLWPyGDwAABQ\nnzb5HCEhZP/+/VVVVYmJicuWLVN4hs/Ly2vHjh2yt1FRUaNGjTp9+vTp06dlhTo6Ovv27Xvb\nK4rh4eHXr19PS0sbPHiwrFBfXz86Orp2U0ZGRu7u7mfPnpXPkatXrz5x4sTt27cNDQ3Ly8vN\nzMwIIdHR0dHR0ZcvX679BAWDwQMAQJ3aZI+QEMLj8U6dOnX06NFx48Z16tRJQ0PDyMjogw8+\nOH78+NGjR7W0tGQ17e3tr1+//vHHHxsZGWlqapqYmHz88cfXr1+XT2ZK6tSp05UrVxYtWmRh\nYcHlcs3NzWfNmnXz5s0BAwbUWf/w4cO+vr6mpqYCgWDYsGEJCQmrVq1atWpVx44dTUxMCCGD\nBg2aPXt2hw4d6jsig8EDAEDd3sHiv9CetNpVyFtnYK0zqhaiVh8W2pO22iMEAABgBBIhAACo\nNSRCAABQa0iEAACg1pAIAQBArSERAgCAWkMiBAAAtYZECAAAag2JEAAA1BoSIQAAqDUkQgAA\nUGtIhAAAoNaQCAEAQK0hEQIAgFprqwvzAkBziESisLAwa2trNpuxb8NpaWmynwyiKCo9PT0w\nMJDL5TLbMgANiRBAHa1fv37VqlUt0XJ4eHh4eDjjzbLZ7ODgYMabBSBIhADqycrKihASEBDg\n5OTEVJtisTgxMXH8+PEaGkz+Ybl8+fLmzZvpgAFaAhIhgDqir4g6OTlNnTqVwWanT5/OYGsy\nmzdvZvASLoAC/NsCAAC1hkQIAABqDYkQAADUGhIhAACoNSRCAABQa0iEAACg1vD4BLR2Ss6B\nouS0JpimBAAUIBFCa/dWc6AoOa0JpikBABkkQmjtlJwDRclpTTBNCQAoQCKE1k75OVCUnNYE\n05QAgDwkQoAmUubmJe5cyjR5vYsmL2qhDmcVGIFECP/AX/a3ovzNy3d557LVji1q5noXTV7U\n4v79+15eXvVtVYd/qNAoJEL4R3h4OLN/2Xk83rJly5oZFZ/Pl/1sPgZbs7e3J4QsX77c0dGx\nvjpK3rm8efPmunXr6AabSflfInm3v0dlTledap/Dn3766ccff1Ry97i4uLi4uIbrMPIBoe1i\nSaVSVccArYUynQkl/7Iz+EVbIpH89ttv77//PofDaWZTzLbWOptSskfYpn+PDH5G9AiBIBEC\nAICaw9g5AABQa0iEAACg1pAIAQBArWHUqLqoqKgoKipipKmSkhJ9fX1GmiKE7CjkMtXa+dfi\nUR2Z+Sf9IPIPppq6cf75oFHmjDRFCLFirrWa8881W19TzLbGbGD85UOYagpaFQyWaa6hQ4de\nvXq1dvmJEycmT578/Pnzrl27vvuoasvLy3v48KGqo6jDe0IHVYdQB3uPKFWHULcUwlJ1COrL\nULpY1SFAi0CPsLmuXLlCvwgLC4uIiMjNzZVtwpcMAIDWD/cIW0pubi6LxaLz4v79+3v16iUQ\nCAYOHPjbb7/RFWoXZmZmslisvLw8uoK/v/+sWbPo169evfLy8hIIBF26dFm6dGl1dbUKPhIA\nQHuEHmGLe/TokZ+fX1paWq9eveLi4iZOnFhQUPDs2bPahQ004uXlNXjw4L/++uvly5fTpk1j\ns9nr1q17Zx8BAKAdQ4+wxWVmZlIU1aFDBz6f/9lnn5WXl/P5/DoL62shNTU1PT1948aN+vr6\nffr0WbVqVUxMzLv8CAAA7Rh6hC3OxcWlX79+1tbWI0eOHDt27Mcff2xsbFxnYX0t3L59u6Cg\nQGFiKrFY3PDkWAAAoAz0CFucjo5OWlpacnKyo6NjZGSknZ3dy5cv6yxU2LGqqop+IRKJ+vXr\nJ/03ZEEAAEYgEb4LbDZ75MiRYWFh9+/f19TUvHbtWp2FmpqahJCKigp6rwcPHtAvbGxsMjIy\nysrKVBU/AEA7hkTY4iIiImxsbB4/flxdXf3bb7+9efPGwcGhzkJTU1M9Pb39+/dXVFTExMTc\nvHmTbmHcuHHdunVbvHjx69evnz9/PmnSpClTpqj2QwEAtBtIhC1u/vz577333qhRo/T19Rcv\nXnzgwIFu3brVWaipqbl79+6oqKguXbqcOnXK19eXbkFDQ+PUqVNPnjwxMTEZMGCAgYFBVFQr\nfdwbAKDNwcwy6iL32M2Mjq3x4ipmlnkrmFlGhTCzTHuFARf/cv369ZCQkEuXLonF4p49ey5Y\nsMDHx6f564iGh4efPn369OnTubm55ubm9Lxr3bp1CwsL8/b2li9k5FPUqXMln++V2nLtN1kh\n+UPVIdThPVUHUJ/3SGv85or0DG0aLo3+IyUlZdSoUUOGDHn06FFeXt769evXrl27eDGT3wG7\ndu0qlUoVEl6dhQAA8G4gEf5NIpH4+Pj4+/uvXLmyS5cuAoHA3d199+7dW7duzc/PJ4RkZGS4\nubnx+fxevXpt2rSJvqRMz6MWGxtrZWXF4/FcXV1lKzxcvXrVwcGBx+MNGTJENtu1bN41Gxub\nZ8+eTZ8+/dtvv5WfjO1tjwIAAM2ERPi3a9euZWdn+/j4yBeOHTtWIpEYGRnV1NSMGzeuf//+\nf/3118mTJ7dt27Z161ZZtW3btv36669ZWVk5OTnh4eGEkKqqqokTJ3p4eLx8+XLlypX79+9X\nONyjR48sLS0PHToUHBwsK3zbowAAQPMhEf4tKyuLw+H06tWrzq3x8fFVVVVhYWEdO3bs27dv\nUFDQpk2bZFvXrVtnaWlpZmY2efLke/fuEULi4uJ4PN7atWsNDAzGjx8/Y8YMZWJ426MAAEDz\nIRH+jR4RQ1FUnVvv3r07ZMgQ2aiZvn375uTkCIVC+m3Pnj3pF3w+n54O5u7du/b29mz236e3\nX79+ysTwtkcBAIDmQyL8W48ePSQSyZMnT+QLnz9/zmKx/vjjD6FQKD+lWVVVlYaGBpfLpd/K\nEp5MTU2NfKFEIlEmhrc9CgAANB/+tv5t0KBB5ubmu3btki+MioqysLDo379/7969//zzT1n5\n5cuXbW1tG8hM1tbWf/75p+wZzT/+UOoJgbc9CgAANB/+yP6Nw+FERERs2bIlPDy8qKiosrIy\nKipq/fr1kZGRbDZ7xowZJSUla9asKSkpuXLlyubNmwMCAhpobebMmeXl5SEhISUlJSdPnjx5\n8mTtOjweT2ENwrc9CgAANB8S4T8+/PDD33///cyZMzY2Np07d46JiTl16tTYsWMJIbq6uomJ\niadOnTIyMpo6deqiRYs+++yzBprS1dX96aef4uPjjY2N//vf/3799de168yYMWPJkiXffvut\n/F5vdRQAAGg+TLGmLq4uSO4VeUfVUbQZrXMCl1ZLTWaWwRRr7RWmWPuXQYMGydZ8IIR07Nhx\n7Nix4eHhnTt3ftumCgsLO3fu/OTJEw0NjabNoCabg+1tD12nguFm8yJvM9JUq/2rx2D2YvAz\nMptTW21gDGIwsDtJs5lqipDWer6g2XBpVNHy5ctli9/+8ccff/7557x585rTIGZQAwBozZAI\nG2Jpaenj45OcnEy/ffXqlZeXl0Ag6NKly9KlS6urqwkhN27c0NHRiYmJsbCw6Ny581dffSUS\nieQbkZ9B7cGDB6NGjeLz+b179z5y5Ahd4eeffx4wYACfz+/SpYuPjw/drPwcbPUdGgAAmg+J\nsHFaWlr0Cy8vLwsLi7/++is1NfXMmTMrV66kyysrKw8dOnTz5s1r166lpKSEhobW2Y5IJPLw\n8HB2di4sLFy3bt2nn36akZEhkUhmzJjx1VdfFRcX37hx48qVK/RagwpzsNV3aAAAaCYkwnpR\nFHXv3r3IyEh66GZqamp6evrGjRv19fX79OmzatWqmJgYuqZUKg0LC+vcuXP37t1Xrlyp8DCi\nTHx8fEVFxZo1awQCweTJk//zn/+cPXuWw+GUlpbOmjVLJBK9fPlSLBbn5OQo7NjAoQEAoJkw\nWEbRunXr1q1bJ3trbGy8ZMkSQsjt27cLCgoU1iYUi8WEEDabbW9vT5fY29vn5+eXl5fXbvn+\n/fsDBw6UtfD777/TLyIjI9evX19eXt6nT5+qqqra43jrO7T8NDQAANA06BEqkg2WoSjqyZMn\nenp6X375JSFEJBL169dP+m90KtLQ0JDN/0LPAspi1TG0r6KiQlNTU6Hw2rVr/v7+kZGRr169\nSk1NdXR0rL1jA4cGAIBmQiKsF4vF6tat2+TJkzMyMgghNjY2GRkZZWVltWuKRKLs7Gz69a1b\nt8zNzQUCQe1q1tbWDx48kL0dMmTI7t27b9y40b17d1dXV7pQtnKhvAYODQAAzYRE2AgDA4Oc\nnJzi4uJx48Z169Zt8eLFr1+/fv78+aRJk6ZMmSKrtmTJkqKiort374aGhiosaihDz6C2cePG\nsrKyqKio9PT08ePH9+nT59mzZ1euXCkuLl6yZElGRkZFRQW9CIZsDraGDw0AAM2BRNiIzp07\n0/OFamhonDp16smTJyYmJgMGDDAwMKCHd9KcnJxsbGxGjx7t6elZ54RqhBA9Pb2ffvpp//79\nXbp02bZt28mTJ01MTEaPHr1o0aJx48b16dOHw+Hs27dvz549P//8M5Gbg63hQwMAQHNgirXm\nunHjxuDBg1v/aUw88CDo0yRGmsLMMm9FHWaWYfafROudWWZcRwZbg9YDibC52koiBACAOuHS\nKAAAqDX0CAEAQK2hRwgAAGoNiRAAANQaEiEAAKg1JEIAAFBrmK9SXSRdS1kZvZGRph6+ed7H\nohcjTRFCJo8d0sesJyNNlde80dcyYKQpESXU5+oz0lR+Vb4uV4+Rpgghjp2H6DEUGIsQNouZ\nr8KFwlcshpoihBAi1WRzGWlIQok12IoT/DZZJ54xU01Bq4JEqC6KyopvZtxhpq3O/JtP7jHT\nFCHzDF0LhQWMNKXB1mCqqQ6aguLq14w0VUPVFAkLGWmKENJBswMllTDSFIfFoaQUI03VUKLG\nKymNy9FiqkEWYTEbG7RLuDQKAABqDYkQAADUGhIhAACoNSRCAABQa0iEAADN8vTpU1Y96CXV\n5s6dK1/IZrM7duzo4eHxxx9/EEIiIiLq3DclJYUQcuzYMYXy1atXKwQwadKkBQsWvPPP3X4g\nEQIANEu3bt2k/5aVlaWvr29tbT1y5Ei6jqGhoWyrRCK5fft2RUXFe++9V1RU5O/vr7C7p6en\njY2Ns7MzISQzM9PNzU1+qywRisXizMzMsLCwU6dOqeqztw9IhAAATBIKhV5eXjU1NcePH+/Q\noUPtCiwWy9zc/Ouvvy4tLU1NTVXYGhkZmZCQEBUVxeVyCSGZmZndu3ev80BhYWFWVlYrVqyg\nKGYeg1FbSIQAAExauHDhrVu39uzZY2tr22hlPp8v/zY/Pz8wMHD+/PlOTk50SWZmZo8ePerc\nNzg4mO4jmpmZNT9sdYZE2A6JRKLvv/9eJMJzxADvWkxMzJ49e7766itvb+/66kgkkvv374eE\nhNjY2Lz//vvym4KDgymKkr8LmJmZ+csvv1hZWQkEAjs7ux07dqD/xzjMLNMOrV+//ptvvpFI\nJMHBwaqOBUCN3L1719fXd9iwYRs2bFDYVFRUxGKx5Eu4XG5ycrKWlpasJCcnJzo6OiQkxMjI\niC6prq7+66+/Bg8evH//fn19/fj4+AULFmRlZW3cyMx0iUBDj7AdsrKykv0EgHejrKxsypQp\nurq6R48e1dRUnOBUYbDM48ePe/bs6enpKX/lJiIigsfj+fv7y0q0tLQoijp+/LiFhYWuru7s\n2bPnzZu3devWqqqqd/Sp1AMSYTvEZrNlPwHg3Zg9e3Z2dvbhw4dNTU0brslms62srObNm1dU\nVPTs2TO6UCwWx8TEeHt71zm+RsbW1lYkElVWVjIWNyARAgA03+bNm48dO/b999+7uLgouYuJ\niQkhpKKign578eLF/Px8Ly8v+TqxsbEsFisnJ0dWcuPGDXNzc0NDQ2biBkIIEiEAQDOlpaUt\nX7588uTJS5cuVX4vuudHP1NPCDlz5oympuaIESPk63h4eHTt2nXOnDnp6emlpaW7d+8+cOBA\naGgog8EDQSIEAGimHTt21NTUnDhxovbsMImJifXt1aVLF0LIli1b6LdXr151cHBQeJrCwMAg\nJSXFwMDA2dnZ2Nh427ZtkZGRn332WYt+HDWEUaMAAM1y4MCBAwcONFBhz549e/bsUSh0dHSU\nSqWyt/SEarVZWVnFxcU1HEBubq5ykULd0CMEAAC1hkQIAABqDYkQAADUGu4RKkskEoWFhVlb\nW7f+5/PS0tIIIWKxWNWBAAC0AUiEylq/fv2qVatUHcVbSEpKmj59uuztpfSbKgwGAKDVQiJU\nFj1jWUBAgGxW+FYrLS0tPDzcw8NDvnC4/ZDIS0eZOUABk9M7iSSM9VwFmtpMNVVDtf/+tMLU\nl60Hm7lbNhKphKmmoB1DIlQWfUXUyclp6tSpqo6lceHh4Roa+OUCADSutd/uAgAAaFFIhAAA\noNaQCAEAQK0hEQIAgFpDIgQAaJanT5/Wnm6b9vPPPxNC5s6dK1/IZrM7duzo4eFBLz0RERFR\n57707KPHjh1TKF+9ejV9XJFIFBQUZGlpyeVyTU1N/fz8ZIs6wVtBIgQAaJZu3bpJ/y0rK0tf\nX9/a2nrkyJF0HYUV6m/fvl1RUfHee+8VFRX5+/sr7O7p6WljY+Ps7EwIyczMdHNzk98qS4SL\nFi3at2/foUOHSkpKdu3atX//fl9fX1WdhDYNiRAAgElCodDLy6umpub48eN1LjfPYrHMzc2/\n/vrr0tLS1NRUha2RkZEJCQlRUVFcLpcQkpmZ2b1799qNlJeX7927NygoaNiwYdra2uPHj/fx\n8YmNjUWnsAmQCAEAmLRw4cJbt27t2bPH1ta20coKCxDm5+cHBgbOnz9fNnFHZmZmjx49au+Y\nnZ0tFAodHR1lJfSS9xRFNSt6tYRE2A7R/yfg/weAdy8mJmbPnj1fffWVt7d3fXUkEsn9+/dD\nQkJsbGzef/99+U3BwcEURckufhJCMjMzf/nlFysrK4FAYGdnt2PHDvp/bXt7e6lUOnToUFmb\n8fHx3t7edfZBoWGYfKQdysjIkP0EgHfm7t27vr6+w4YN27Bhg8KmoqIihTntuFxucnKylpaW\nrCQnJyc6OjokJMTIyIguqa6u/uuvvwYPHrx//359ff34+PgFCxZkZWVt3LhRvqkXL17MmTPH\nxsZmx44dLfPJ2jkkwnZo2bJlHA5n8eLFqg4EQI2UlZVNmTJFV1f36NGjmpqaClsNDQ0LCwvp\n1xRFZWVlTZw40dPT88WLF/S9QEJIREQEj8fz9/eX7aWlpSV/aWf27Nl//PHH1q1bQ0ND6Wuq\nFEVt2bLlyJEjISEh7u7uLfsJ2y9cGm2HuFzuihUrZP93AcA7MHv27Ozs7MOHD5uamjZck81m\nW1lZzZs3r6io6NmzZ3ShWCyOiYlp9Nqmra2tSCSqrKwkhLx+/drd3T0vL+/ChQvIgs2BRAgA\n0FybN28+duzY999/7+LiouQu9NgW2SDPixcv5ufne3l5ydeJjY1lsVg5OTmykhs3bpibmxsa\nGhJCZs2apa+vHxYWhhn2mwmJEACgWdLS0pYvXz558uSlS5cqvxfd86OfqSeEnDlzRlNTc8SI\nEfJ1PDw8unbtOmfOnPT09NLS0t27dx84cCA0NJQQcu/evVOnTi1fvpy5z6G+kAgBAJplx44d\nNTU1J06cqD07TGJiYn17denShRCyZcsW+u3Vq1cdHBwUnqYwMDBISUkxMDBwdnY2Njbetm1b\nZGTkZ599Rgg5f/48IWTIkCEKR5TdiQTlsaRSqapjaBuOHj360UcfxcXFtYn1CGs7cCHh0x1v\n8XW1IYwuzLt1E0NREWLA02WqKRZhbNHaGqqGqaYIIVN6TGOqKQ02Y9fT8ipzmWqKEMLj8Buv\npBxmF+Y1E3RjsDVoPdAjBAAAtYZECAAAag2JEAAA1JriTQKRSBQWFmZtbc1mI0f+S1paGiFE\nLBarOhAAAGCSYiJcv379qlWrVBJKm5CUlDR9+nRVR9EUeZUvVR0CAEBrpJgIraysCCEBAQGy\nuc+BlpaWFh4e7uHhoepAmmjS0Pf0zJmZhlskYbJbvHCR4qyMTdZzTOOT/SspZmYgU01VSYRM\nNUUIeVhyh6mm2CzGrvpUS6qZaooQwmdu1KiUMDkqHqNG2yvFREhfEXVycmqjDwm0qPDwcMzg\nAADQzuBGIAAAqDUkQgAAUGtIhAAAoNaQCAEAmsXFxaX2LKM0hRV0y8vLe/fuPXPmTPptRERE\nnXulpKQQQsaOHStfqKen5+bmRm+iiUSioKAgS0tLLpdramrq5+cnW8sC3goSIQBAs6Smpkr/\nx9DQ8OOPP5a9XbJkiXxNX1/fjIwM2Vt/f3/pv3l6etrY2Dg7O9MVHBwc6HKJRHLnzp2BAwe6\nubkdPHiQ3rpo0aJ9+/YdOnSopKRk165d+/fv9/X1fWefuj1BIgQAeBeio6OTk5Otra3rqxAZ\nGZmQkBAVFVV7VW02m21pablu3Tpvb+8FCxYUFxeXl5fv3bs3KCho2LBh2tra48eP9/HxiY2N\nRaewCZAIAQBa3KNHj7788ssDBw506tSpzgr5+fmBgYHz589v+BnuhQsXvnnzJj4+Pjs7WygU\nOjo6yjbRK/1SFDOPC6sVPBXX5tU5Kx6Px/Pw8OBwOCoMDABoQqFw2rRpfn5+bm5uISEhddYJ\nDg6mKGr16tUNN2Vvb08IuX37to+Pj/wiehKJJD4+3tvbm17vF94KEmGbV9+seMnJya6uru8+\nHgBQEBAQIBAI6JXl65STkxMdHR0SEmJkZNRwU9ra2jo6OkLhv6YrevHixZw5c2xsbHbs2MFM\nxGoGibDNq3NWPB6PN3r0aNUFBQB/O378+JEjR27dutXAvFQRERE8Hs/f37/R1sRicWVlJb26\nPSGEoqgtW7YcOXIkJCTE3d2dsaDVDBJhm4dZ8QBas7NnzxYXF3fr1k1WcunSpdjYWNk1G7FY\nHBMTo+RVzZs3b1IUNXToUELI69evP/roo0GDBl24cAGzPzYHBssAALSgnTt3yj8gMWrUKPr5\nCtmdi4sXL+bn53t5eSnT2pYtW0xMTMaMGUMImTVrlr6+flhYGLJgM+H0AQCo0pkzZzQ1NUeM\nGNFAHYlEkpOTs3nz5qNHj544cYLL5d67d+/UqVPXrl17Z3G2Y0iEAACqdPXqVQcHBz6/jsWn\nbt++zWKx6NcdO3YcMmTIuXPnhg8fTgg5f/48IWTIkCEKuxQUFNT3hAbUB4kQAIAxhYWFDVdI\nTU1VKJGfNU3e6dOnG2jHz8/Pz8/vbUKDeuEeIQAAqDUkQgAAUGtIhAAAoNZwj1BZ9Ax+ly9f\nVnUgitLS0gghYrFY1YEAALRJSITKohdP2bx58+bNm1UdSx2SkpKmT5/eQIVXlXnvLBgAgDYE\niVBZy5YtoyhKYW7r1iAtLS08PNzDw6PhasbaJo9LHzFyRIGmNiPt0HqOsWWqqawzD5hqqmZG\nDVNNvRG9YaopQoi2Rh2D7JtGk63JVFMVNZVMNUUIEYqFjVdSjhZHi6mmoB1DIlQWl8tduXKl\nqqOoW3h4OKaWAABomtbVuQEAAHjHkAgBAECtIRECAIBaQyIEAAC1hkQIANAsLi4urHps3LhR\nvmZ5eXnv3r1nzpxJv42IiKhzL3r20bFjx8oX6unpubm51Tkx6aRJkxYsWPAOPml7hUQIANAs\nqampsuUGDQ0N6eUGaUuWLJGv6evrSz+RTPP395f+m6enp42NjbOzM13BwcGBLpdIJHfu3Bk4\ncKCbm9vBgwfprWKxODMzMyws7NSpU+/sw7ZLGHMPAPAuREdHJycnW1tb11chMjIyISHh4sWL\nXC5XYRObzba0tFy3bl1ubu6CBQvGjRtnYGAQFhb2zTfftHDUagE9QgCAFvfo0aMvv/zywIED\n9S0WmJ+fHxgYOH/+fCcnpwbaWbhw4Zs3b+Lj4wkhwcHBdH/RzMys/zdLUQAAIABJREFURYJW\nG+gRtnl1ToLK4/E8PDw4HI6KggKAfwiFwmnTpvn5+bm5uYWEhNRZJzg4mKKo1atXN9yUvb09\nIeT27duMB6nOkAjbvPomQU1OTnZ1dVVRUADwj4CAAIFAEBoaWl+FnJyc6OjokJAQIyOjhpvS\n1tbW0dERChmbhQ4IEmE7UOckqDweb/To0SqMCgBox48fP3LkyK1btxqYBzEiIoLH4/n7+zfa\nmlgsrqys7NKlC6MxqjskwjavNU+CCgBnz54tLi7u1q2brOTSpUuxsbGyazZisTgmJsbb27tD\nhw6Ntnbz5k2KooYOHdpyAashDJYBAGhBO3fulH9AYtSoUfTzFbI7FxcvXszPz/fy8lKmtS1b\ntpiYmIwZM6YlQ1Y76BECAKjSmTNnNDU1R4wY0UAdiUSSk5OzefPmo0ePnjhxovbzFdAcSIQA\nAKp09epVBwcHPr+OlSZv377NYrHo1x07dhwyZMi5c+eGDx/+bgNs/5AIAQAYU1hY2HCF1NRU\nhZI6Z00jhJw+fVrJg+bm5ipZE+qEe4QAAKDWkAgBAECtIRECAIBaQyIEAAC1hkSoLgqF+aoO\nAQCgNcKoUXVhom36vPwpI03VUGJG2qHFzAxkqqmaGTVMNTX6szlMNcWyNWCqKUJI7MIljVdS\nTpWYsfkq+Ro8pppiFoOfkRAy3Pg9BluD1gM9QgAAUGtIhAAAoNaQCAEAQK0pJkJ6lVf6JwAA\nQLunmAgfPHgg+wkAAMqws7NjyREIBC4uLjdv3qS3uri4sOqxceNG+XbKy8t79+49c+ZM+m1E\nRESdeynMyqawFyFEJBIFBQVZWlpyuVxTU1M/P7+KiooWPgdtmGIitLW1lf0EAAAlTZkyhV5o\nSSwW3717VyKRuLu7l5WVEUJSU1NlyzAZGhrSyzDRliz51zBgX1/fjIwM2Vt/f3/pv3l6etrY\n2Dg7OzewFyFk0aJF+/btO3ToUElJya5du/bv3+/r69uSn75tU0yE9Crn8mudAwCA8jgcTo8e\nPVauXPn69evz588rv2N0dHRycrK1tXV9FSIjIxMSEqKiouSXYaq9V3l5+d69e4OCgoYNG6at\nrT1+/HgfH5/Y2Fh0CuuDhAcAwDyJREIIEQgEStZ/9OjRl19+eeDAgU6dOtVZIT8/PzAwcP78\n+U5OTg3vlZ2dLRQKHR0dZSUmJiYEgz/qh0TY3ohEojVr1pw8eZL+/xAA3jGJRJKenr5q1aoB\nAwaMGjVKmV2EQuG0adP8/Pzc3NzqqxMcHExR1OrVqxvdy97eXiqVDh06VBZPfHy8t7d3hw4d\nmvJ51ABmlmlv1q9fv2rVKkJIcnKyq6urqsMBUBfHjh2TLaJLCOFwODt37uRwOMrsGxAQIBAI\nQkND66uQk5MTHR0dEhJiZGSk/F6EkBcvXsyZM8fGxmbHjh3KRKKekAjbGysrK0LI8uXLR48e\nrepYANTIlClT4uPj6dfFxcXx8fG+vr4lJSUKw2FqO378+JEjR27duqWhUe8f5IiICB6P5+/v\nr/xeFEVt2bLlyJEjISEh7u7ub/+B1AgujbY39EAnR0dHJb+KAgDjDAwMfHx8xo0bFxMT02jl\ns2fPFhcXd+vWjX404tKlS7GxsSwW69dff6UriMXimJgYhWubDe/1+vVrd3f3vLy8CxcuIAs2\nCokQAKBFVFVVyQ/vrM/OnTvlH5AYNWoU/XyF7NbGxYsX8/Pzvby8lN9r1qxZ+vr6YWFhDfQy\nQQbnCACAYQUFBT/++GNKSkpUVFTzWztz5oympuaIESOUrH/v3r1Tp05du3at+YdWE0iEAAAM\nkB8sw+fz+/bte+DAgRkzZjS/5atXrzo4OPD5fCXr0w8vDhkyRKG8oKCgvmcz1BwSIQBAc927\nd0/JmoWFhQ1XSE1NVShRmFCt0b38/Pz8/PyUjAcI7hECAICaQyIEAAC1hkQIAABqDYkQAADU\nGgbLvB2RSBQWFmZtbd1qF+hIS0sjhIjFYlUHAgDQNiARvh3ZTJ6tXFJS0vTp0+VL8qvyVBUM\nAEBrhkT4duiZPAMCAuRXQmlV0tLSwsPDPTw8FMo78jo/KL6vkpAaViURMtXUG9Ebpppi2Row\n1ZT0QTFTTRFCymsYW1JOKK5mqqlqiYippgghEuZWC5JIsQYLNA6J8O3QV0SdnJymTp2q6ljq\nFR4ejnmVAACU1EpvdAEAALwbSIQAAKDWkAgBAECtIRECADSXnZ0dS45AIHBxcbl58ya91cXF\nhVWPjRs3yrdTXl7eu3fvmTNn0m8jIiLq3Eth9lGFveRNmjRpwYIFLfOh2w8kQgAABkyZMoVe\nGlAsFt+9e1cikbi7u5eVlRFCUlNTZQsHGhoa0gsH0hTWr/f19c3IyJC99ff3l/6bp6enjY2N\ns7NzA3sRQsRicWZmZlhY2KlTp1ryQ7cTSIQAAEzicDg9evRYuXLl69ev6RWRlBQdHZ2cnGxt\nbV1fhcjIyISEhKioKPn1fuvcKywszMrKasWKFRRzz6K0Y0iEAADMk0gkhBCBQKBk/UePHn35\n5ZcHDhyob8nA/Pz8wMDA+fPnyz/EXN9ewcHBdA/SzMysqZ9AjSARtjf0F8CbN2/S/x8CwDsm\nkUjS09NXrVo1YMCAUaNGKbOLUCicNm2an5+fm5tbfXWCg4Mpilq9evVb7QXKwGPX7Q19q2Dd\nunWurq6urq6qDgdAXcivUE8I4XA4O3fu5HA4yuwbEBAgEAhCQ0Prq5CTkxMdHR0SEmJkZKT8\nXqAkJML2ZtmyZRRFOTg4jB49WtWxAKiRKVOmxMfH06+Li4vj4+N9fX1LSkoUhsPUdvz48SNH\njty6dauBCaEiIiJ4PJ6/v/9b7QVKwqXR9obL5a5cuXLixIlKfhUFAMYZGBj4+PiMGzcuJiam\n0cpnz54tLi7u1q0b/WjEpUuXYmNjWSzWr7/+SlcQi8UxMTHe3t4dOnRQfi9QHhIhAECLqKqq\nkh/eWZ+dO3fKPyAxatQo+vkK2a2Nixcv5ufne3l5vdVeoDz0qQEAGFZQUPDjjz+mpKRERUU1\nv7UzZ85oamqOGDGi+U1BnZAIAQAYID9Yhs/n9+3b98CBAzNmzGh+y1evXnVwcODz+c1vCuqE\nRAgA0Fz37t1TsmZhYWHDFVJTUxVKFCZUU3IvWm5urlJhqTfcIwQAALWGRAgAAGoNiRAAANQa\nEiEAAKg1JEJ1IZFi6lEAgDpg1Ki66NGhp6m2qaqjqMPDkjtMNaWtwdj48tiFjUyLpbzymgqm\nmiKEzPvyO8ba6qXLVEsfjhvGVFOEEEoqZaopCXNNEUK+smewMWhF0CMEAAC1hkQIAABqDYkQ\nAADUGhIhAACoNSRCAIDmsrOzY7FYgYGBCuUnTpygl0kSCoV0iVAoXL9+/YABAwQCgUAgGDJk\nyP79+2X1vb29WXW5cuVKREREnZvqnIBNJBIFBQVZWlpyuVxTU1M/P7+KCiYHbbUzSIQAAAzg\ncDiHDh2S/nuc6qFDh+RXBi0vL3dxcdm+ffvKlSvz8vKePn06d+7cL774Yvny5bI6ZmZm0lqG\nDh3q7++vUOjp6WljY+Ps7Fw7mEWLFu3bt+/QoUMlJSW7du3av3+/r69vy332tg6JEACAAaNH\nj87Jybl06ZKspLy8PDEx0cXFRVaydOnShw8fnj9/fvLkyR06dOjcufO8efM2bty4YcOG69ev\nv9XhIiMjExISoqKi6CUP9fX1V65c6e7urqWl9fTp07179wYFBQ0bNkxbW3v8+PE+Pj6xsbHo\nFNYHiRAAgAEWFhb/+c9/Dh48KCtJSEjgcDgeHh702/Ly8qioqE/+v717j4uqTPwH/gw4AwPO\nhKCSTEi5maz5giXzmn4FFLfQflyEuHjdFFHB9qWpTOHipZdJgmHGJm7ExYoRRYXVcgGVCFTU\n1HJpzdxaIyFkMRAHGMa5/P44u/OaHYgG5pE5zPN5/9GLOWfmM8+ZRj6cmXPOs3ixl5eX8QOX\nLl0qFosLCwvNf66mpia5XB4fHz9t2jTDwp07d0ZFRbW0tLS1talUqokTJxpWjRo1ihCi0+n6\nt2k2D0Voa9Rq9fbt20tKSrRaXEoGYOCo1erY2NjDhw9rNBpuiUKhCA0NdXBw4G5evXpVrVZP\nn2568QGxWNze3p6enm7+c23evFmn023dutV4YVhY2PLly52cnHx8fLhPU7nlWq22qKgoOjpa\nIpH0b9NsHorQ1uzatWvLli2hoaEVFRXWHgsAQ/R6PbdDVlZWRgi5e/dueXl5TEyM4VvDtrY2\nQoirq2vvOfX19SaHw5h0ZF1dXV5enlwuHzlypPFyPz+/7mkNDQ3z58/39vZ+//33Ldk622Z6\niTVu3/n8+fPWGMwgcO7cOUKI4S8+Hho7diwhJCkpKSAgwNpjAWCLu7t7YGBgQUFBcHBwUVGR\nVCoNCgrav38/t5arwMbGxt5DZDJZ77PpZmZmOjo6JiYmmiwXCoXGN3U63d69ewsLC7dt2zZ3\n7tw+bwxLTIvw5s2bhJCMjIyMjAxrjGdw+PTTT2NiYqw9ip7Z2dkRQiZOnGh8rBoADIzY2Ni1\na9d2dHQoFIrIyEjjcvLx8RGJRFeuXFm2bJnxQ/R6/WOPPTZ//nxDZfZCo9Hk5+f/6uecP//8\n80svvfTss89WVVUNGYJrSv8K0xdo06ZNOp1u3Lhx3O9TMHHu3Lk9e/YYvv0GADAWHh6+evXq\nffv2VVVVbd++3XiVs7NzTExMTk5OcnKyu7u7YfmhQ4caGhrM/Nu6urq6qakpIiKi97stW7bM\nxcUlNTW1H5vAINMiFIlEKSkpVhnKYLFnzx78hQUAPZJKpfPmzUtOTvbw8Jg5c6bJ2t27d1+6\ndGn27NkZGRnPPfecRqM5evToK6+8kpSUZHyWRS9KS0uFQmH3ZGO1tbXHjx+/ePFiv7eCNdjt\nAwCgaeHChV1dXVFRUQKBwGSVm5tbTU1NSEhIQkKCi4uLp6dnbm5udna2+btuFy5c8PX1FYt7\nm3GssrKSEDJ58mSTg26am5v7t0U2D3s2AACWqq2tNfwcFhZmfH2ZxMRE4wNbJBLJjh07duzY\n0WPOwYMHe3+iHi+oRghpbW01/JyQkJCQkGDOsIGDPUIAAGAaihAAAJiGIgQAAKahCAEAgGk4\nWIbv1Gp1amqq+Wd28v/aNwAAvIIi5Dvu2qF9fVT3a9/o9LgGNwBAD1CEfMddO3TdunXG8630\n4peufWNvN8SeUhd2PzvKEnYCap/PC+2Ev34n83RqVLSiVJouWlGEEPKklFrUP9toJXU8eEAr\nihCi//W7mEuLiYfADChCvuM+EZ02bVpkZKSZD8G1bwAAzIeDZQAAgGkoQgAAYBqKEAAAmIYi\nBACw1IQJEwQCgVwuN1l+7Ngx7oLXKtV/Dr9SqVS7du3y8/NzdnZ2dnaePHnygQMHDPePjo4W\n9KSmpiYzM7PHVb909VFOaGjoqlWrqG+vjUERAgBQYG9vr1AojC+3TQhRKBTGU2QrlUp/f//3\n3nsvJSWlsbHx1q1bK1asWLNmTVJSkuE+MplM383UqVMTExNNFoaHh3t7e8+YMaP7YDQazT//\n+c/U1NTjx48/vE22GTi2EACAgoCAgFOnTp09e9bQTEql8sSJE/7+/qdPn+aWbNy48fr169eu\nXfPy8iKESCSSlStX6nS6NWvWRERETJo0yfyn279/f3FxcXV1tUgkIoS4uLi88sorNTU1lZWV\n9fX1WVlZf/rTn2hvos3CHiEAAAWjR4+eMmVKQUGBYUlxcbG9vb3hpF6lUpmTk7N48WKuBQ2W\nLl0qFosLCwvNf66mpia5XB4fH298evHOnTujoqJaWlqGDx++efNmbq9RJpNZtllMQBHaGp1O\nRwi5fPmyVotLyQAMHLVaHRsbe/jwYcMFDhUKRWhoqIODA3fz6tWrarV6+vTpJg8Ui8Xt7e3p\n6enmP9fmzZt1Ot3WrVuNF4aFhS1fvtzJyan/28AqFKGtuXnzJiHkrbfeqqiosPZYABii1+u5\nHbKysjJCyN27d8vLy2NiYgzfGra1tRFCXF1de8+pr683ORzGpCPr6ury8vLkcvnIkSONl/v5\n+dHcHpbgO0Jbs2nTJp1O5+vrGxAQYO2xALDF3d09MDCwoKAgODi4qKhIKpUGBQXt37+fW8tV\nYGNjY+8hMpns9u3bvdwhMzPT0dHReNZ7jlBI7RKDrMEeoa0RiUQpKSkhISHGx6oBwMCIjY0t\nKSnp6OhQKBSRkZHG5eTj4yMSia5cuWLyEO6bvPj4eHPyNRpNfn5+dHS0RCKhOW62oQgBAKgJ\nDw/XaDT79u2rqqoymQHG2dk5JiYmJyfnzp07xssPHTrU0NBgcudfUl1d3dTUFBERQXPQzEMR\nAgBQI5VK582bl5yc7OHhMXPmTJO1u3fv9vLymj17dnl5eUdHR1tbW15eXlxcXFJSkr+/vzn5\npaWlQqGwezJYAkUIAEDTwoULu7q6oqKiuk9Y5ubmVlNTExISkpCQ4OLi4unpmZubm52dnZqa\namb4hQsXfH19xWIx7VEzDQfLAABYqra21vBzWFiY8fVlEhMTjQ9skUgkO3bs2LFjR485Bw8e\n7P2JfumCaq2trT0u7/24G+BgjxAAAJiGIgQAAKahCAEAgGkoQgAAYBqKkBUqbae1hwAAwEc4\napQVer3+jqre2qPoQZe2i1ZU+4MOWlHiIY60orq0alpRhJD/94LpJZv7rePBA1pRp7JO04oi\nhAyZMJxWlEatoRVFCCGRNMOAP1CEfcPN7XD+/PkBe8Zz584RQgzXswcAALpQhH3Dze2QkZGR\nkZExkM/76aefmnkFJgAA6BMUYd9wczuMGzfOzm6Avl49d+7cnj17DHN7AgAAXSjCvuHmdhjg\nJ92zZ8+QIfg/BQDwUOCoUQAAYBqKEADAUhMmTBAIBHK53GT5sWPHuFnmVSoVt0SlUu3atcvP\nz8/Z2dnZ2Xny5MkHDhww3D86OlrQk5qamszMzB5XcVcfVavVycnJXl5eIpHIw8MjISGhvb19\nwDZ/sEMRAgBQYG9vr1AojC+3TQhRKBTGU2QrlUp/f//33nsvJSWlsbHx1q1bK1asWLNmTVJS\nkuE+MplM383UqVMTExNNFoaHh3t7e8+YMYMQsn79+tzcXIVC0dra+pe//OXAgQOrV68esG0f\n7FCEAAAUBAQE1NXVnT171rBEqVSeOHHCeKLBjRs3Xr9+vbKyMiwsTCKRjBgxYuXKlenp6Wlp\naZcuXerT0+3fv7+4uDgnJ0ckEimVyg8++CA5OXn69OlOTk7z58+Pi4v7+OOPsVNoJhQhAAAF\no0ePnjJlSkFBgWFJcXGxvb294ZBvpVKZk5OzePFiLy8v4wcuXbpULBYXFhaa/1xNTU1yuTw+\nPn7atGmEkO+//16lUk2cONFwh1GjRpH/nvcMvwpFaGvUavX27dtLSkq0Wq21xwLAELVaHRsb\ne/jwYcPlLxQKRWhoqIODA3fz6tWrarV6+nTTawOJxeL29vb09HTzn2vz5s06nW7r1q3cTR8f\nH+7jU+6mVqstKiqKjo6WSCSWbBE7UIS2ZteuXVu2bAkNDa2oqLD2WAAYotfro6KiWlpaysrK\nCCF3794tLy+PiYkxfGvY1tZGCHF1de09p76+3uRwGJOOrKury8vLk8vlI0eO7P7whoaG+fPn\ne3t7v//++3Q2jAEoQlszduxYQkhSUlJAQIC1xwLAFnd398DAQO7T0aKiIqlUGhQUZFjLVWBj\nY2PvId0PltmwYYPxHTIzMx0dHY1nvefodLo9e/YsWLBg3bp1+fn5Tk5OdLaKAShCW8Nd8mbi\nxInGx6oBwMCIjY0tKSnp6OhQKBSRkZFCodCwysfHRyQSXblyxeQher1eJpPFx8ebk6/RaPLz\n87t/7Pnzzz/PnTu3sbGxqqpq7ty5lm8IU1CEAADUhIeHazSaffv2VVVVmVwf2NnZOSYmJicn\n586dO8bLDx061NDQYObFhKurq5uamiIiIkyWL1u2zMXFJTU1FVeh6gcUIQAANVKpdN68ecnJ\nyR4eHjNnzjRZu3v3bi8vr9mzZ5eXl3d0dLS1teXl5cXFxSUlJRmfZdGL0tJSoVBoklxbW3v8\n+HHjkxGhT1CEAAA0LVy4sKurKyoqSiAQmKxyc3OrqakJCQlJSEhwcXHx9PTMzc3Nzs5OTU01\nM/zChQu+vr5isdh4YWVlJSFk8uTJJkfZNDc3U9kim4edaAAAS9XW1hp+DgsLM76+TGJiovGB\nLRKJZMeOHTt27Ogx5+DBg70/EXdBNRMJCQkJCQl9GzEYwR4hAAAwDUUIAABMQxECAADTUIQA\nAMA0FCEr1Nouaw8BAICPcNQoK8RDnFS6DipRdlT/fhLbi3/9TuZRaVS0oijSUp0BQPe/091Z\ngloQIUMmDKcXRjS19A76H+dCLQpsF/YIAQCAaShCAABgGooQAACYhu8I+Y6bY/r8+fNm3v/c\nuXOEEMPUoAAA0DsUId/dvHmTEJKRkZGRkWH+oz799FMzL2YPAJabMGHC119/nZSUZHLV0GPH\njoWHhxNCOjs7HR0dCSEqlWrv3r0KheLbb78lhDz99NOJiYlLlizh7h8dHV1YWNg9//z58198\n8cXatWu7rzp9+nRgYKDhZmho6KOPPpqVlUVv42wfipDvNm3apNPpxo0bx000+KvOnTu3Z8+e\n4ODghz0wADBmb2+vUCh27txpfK1thUJhb2+v1Wq5m0qlcs6cOY2NjRkZGXPmzFGpVMeOHVuz\nZs3XX3/91ltvcfeRyWS3b9/unj916lSTyXgXLFjwj3/8Y8aMGYQQjUZz69atoqKi48ePx8XF\nPayNtFEoQr4TiUQpKSl9esiePXswJxnAAAsICDh16tTZs2e5ZiKEKJXKEydO+Pv7nz59mluy\ncePG69evX7t2zcvLixAikUhWrlyp0+nWrFkTERExadIk859u//79xcXF1dXVIpGIEJKamvqn\nP/2J9jaxAgfLAABQMHr06ClTphQUFBiWFBcX29vbGz6eUSqVOTk5ixcv5lrQYOnSpWKxuMdP\nRH9JU1OTXC6Pj4+fNm0at2Tz5s16vZ6b7N7iTWEOitDWcIfJXLp0yfBpDAAMALVaHRsbe/jw\nYcOhagqFIjQ01MHBgbt59epVtVo9ffp0kweKxeL29vb09HTzn2vz5s06nW7r1q00Bg4oQptT\nUlJCCElLS6uoqLD2WAAYotfro6KiWlpaysrKCCF3794tLy+PiYkxzE3Y1tZGCHF1de09p76+\n3mR+XZOOrKury8vLk8vlI0eOfDibwhwUoa0ZNWoUISQkJCQgIMDaYwFgi7u7e2BgIPfpaFFR\nkVQqDQoKMqzlKrCxsbH3EJlMpv9fGzZsML5DZmamo6OjyYEzYAkUoa3hjlgbM2aMvb29tccC\nwJzY2NiSkpKOjg6FQhEZGSkUCg2rfHx8RCLRlStXTB7CfbEXHx9vTr5Go8nPz4+OjpZIJDTH\nzTYUIQAANeHh4RqNZt++fVVVVSbn8jo7O8fExOTk5Ny5c8d4+aFDhxoaGsw88be6urqpqSki\nIoLmoJmHIgQAoEYqlc6bNy85OdnDw2PmzJkma3fv3u3l5TV79uzy8vKOjo62tra8vLy4uLik\npCR/f39z8ktLS4VCYfdksASKEACApoULF3Z1dUVFRRmfWc9xc3OrqakJCQlJSEhwcXHx9PTM\nzc3Nzs42uR5NLy5cuODr6ysWU5u8DAghAj296c2AD9avX5+RkbFu3bq3337beHmnpr1FTWea\nN7rzEf7UUU8rqqWrhVbUnc4mWlF3O1tpRRFCSr+/Riuqk94FaSurv6IVRXg8H6E+s5ZiGvAH\n9ggBAIBpKEIAAGAaihAAAJiGIgQAAKahCG0NN5Ev918AAPhVKEJbc+3aNcN/jam1XdYYDgAA\n32HWOluzfPnyioqK5cuXmywX2Tt0aJVUnkKrpzmvhZ5QO4HHwd6BVlSnRkUriu7LpaV3vpOW\n3scGGjW1MzEIoXrOww2a566ArcIeoa3hZunk/gsAAL8KRQgAAExDEQIAANNQhAAAwDQUIQCA\npSZMmCAQCORyucnyY8eOcbPMq1T/OfxKpVLt2rXLz8/P2dnZ2dl58uTJBw4cMNw/Ojpa0JOa\nmprMzMweV505c4YQ0tra+oc//MHNzU0ikUyaNKmoqGjAtt0GoAgBACiwt7dXKBQm0xgoFArj\nKbKVSqW/v/97772XkpLS2Nh469atFStWrFmzJikpyXCf7jPU6/X6qVOnJiYmmiwMDw/39vae\nMWMGIWTx4sVffvnluXPnGhoawsLCXnrppVOnTg3Ytg92KEIAAAoCAgLq6urOnj1rWKJUKk+c\nOGE80eDGjRuvX79eWVkZFhYmkUhGjBixcuXK9PT0tLS0S5cu9enp9u/fX1xcnJOTIxKJbt26\ndeLEibS0tHHjxkkkktdff/3JJ588cuQIrU2zeShCAAAKRo8ePWXKlIKCAsOS4uJie3v74OBg\n7qZSqczJyVm8eLGXl5fxA5cuXSoWiwsLC81/rqamJrlcHh8fP23aNELI119/TQjx8/Mz3KGr\nq6urC9fQMBeK0NZwF1e7fPmyVkvzPG4A6J1arY6NjT18+LDmvxM9KhSK0NBQB4f/XOfh6tWr\narV6+vTpJg8Ui8Xt7e3p6enmP9fmzZt1Ot3WrVu5m/PmzdPr9W5uboQQpVKZmpra2Ni4ZMkS\nC7eIHShCW3Pz5k1CyFtvvVVRUWHtsQAwRK/XR0VFtbS0lJWVEULu3r1bXl4eExNj+Nawra2N\nEOLq6tp7Tn19vcnhMCYdWVdXl5eXJ5fLR44cafLYFStWSCSS1157be7cub/73e+obZutQxHa\nmk2bNm3btq24uDggIMDaYwFgi7u7e2BgIPfpaFFRkVQqDQoKMqzlKrCxsbH3kO4Hy2zYsMH4\nDpmZmY6OjomJid0fm52dfe/evZMnT168eDEkJITCJrEBRWgwHdMdAAAT4ElEQVRrRCJRSkpK\nSEiI8bFqADAwYmNjS0pKOjo6FApFZGSkUCg0rPLx8RGJRFeuXDF5iF6vl8lk8fHx5uRrNJr8\n/Pzo6GiJRGJYuH79+scee4z7WSqVPv/886+88srnn3/e2opLrZoFRQgAQE14eLhGo9m3b19V\nVVVMTIzxKmdn55iYmJycnDt37hgvP3ToUENDg8mdf0l1dXVTU1NERITxQh8fn4aGhoaGBsOS\nrq4uBwcHsVhswaYwBEUIAECNVCqdN29ecnKyh4fHzJkzTdbu3r3by8tr9uzZ5eXlHR0dbW1t\neXl5cXFxSUlJxmdZ9KK0tFQoFJokR0dHy2Sy1atX19fXK5XKoqKijIyMhIQEw3E60DsUIQAA\nTQsXLuzq6oqKihIIBCar3NzcampqQkJCEhISXFxcPD09c3Nzs7OzU1NTzQy/cOGCr6+vya6e\no6NjWVnZgwcPvL29PTw83nzzzd27d6elpdHZHgYI9PSmNwM+69S0t6rvUonS6anNY0cIudP5\nE62oTk0nragbrTdpRd1Xt9OKIoSUfl9LK6rzwQNaUZ99dpVWFCGEmNaHBajOR6gvv00xDfgD\ne4QAAMA0FCEAADANRQgAAExDEQIAANNQhAAAwDQcNQoAAEzDHiEAADANRQgAAExDEQIAANNQ\nhAAAwDQUIQAAMA1FCAAATEMRAgAA01CEQFl7e/v9+/e7L29ubrZiFF38HBg/R0V4PDBjra2t\nXV1d1h4FWAeKEKhRKpVLliyRSqUuLi4LFy7s6OgwXjtixAirRNHFz4Hxc1R8HlhZWdmqVate\nfvnlysrK1tbWWbNmDRs2zMXFZePGjTodzVnGYFDAlWWAmoSEhA8//DAuLu7RRx89fPjwb3/7\n2/z8fMNagaAPbzaKUXTxc2D8HBVvB1ZYWBgTEzN27FiVStXY2BgUFPTFF1+sXbu2q6srIyPj\nnXfeefnllwd+VGBFKELWrVq1qvc7ZGVlmRnl4eHxwQcfvPDCC4QQrVY7f/78devWzZ07l1vb\np996FKMIX7eRn6Pi7cAojmrixIkvvfRSUlISIWTnzp2vv/76yZMnn3/+eUJIUVHRzp07L1++\nbGYU2AYUIevCwsJKSkr0er1MJuvxDrdvmzsrt1AobGtrE4vF3M2GhoYFCxZUVVUNGTKE9PG3\nHsUowtdt5OeoeDswiqOSSCQ//fTT0KFDCSFKpVIikbS1tUkkEu6mTCa7d++emVFgI/TAvDfe\neIPKO8HT0/PcuXPGSzIyMpKSkrif+/QUFKM4PNxG3o6KtwOjNaphw4bduHHDcHP58uWGn3/8\n8UeJRGL5U8DggiIE/Y0bN6j8fnnttddGjRqVmZlZU1PDLdFqtYGBgQsWLLh06VKfnoJiFIeH\n28jbUfF2YLRGtXTpUj8/v/Ly8ubmZuPl9fX1L774YnBwsOVPAYMLihD0Dx48eOONNyzP6erq\nSkhIcHR0NP5t1dzcHBwc3NePHyhGcXi4jbwdFW8HRmtULS0t8+fPNxmATCYTCASjR4/+5ptv\nLH8KGFzwHSFQplQqv/vuO19fX+OFf//73z///POEhARrRdHFz4HdunXr3r17tEbV3Nw8fPhw\nKgPj7ctVW1traMT169f7+vpGR0c7ODhYa0hgLShCMNXZ2dnR0eHm5mb1qJaWlqKioq+++kql\nUrm7u0+dOvX3v/+9SCTiQ5qxU6dOzZkzx/IcC6McHBxaWlqcnJyojIRK2pkzZ2bMmEHlReZn\nFNgIa++SgpVt2bLF8LNWq3311Ve5XxCenp55eXnWitLr9R999BF3IB8hZOjQoXZ2doSQUaNG\nnThxoq9R1NNMUPx3ZElUYGBgTk4OrZFQSSOEeHt7nz9/3vLx8DMKbMOQAW9e4Jdt27Zt3bqV\n+zkrK2vv3r0rV64cP378F198sWLFCq1Wa/7JxRSjTp8+vWTJkujo6LVr1/r4+Dg5OWk0msuX\nL+/duzcsLKyiouK5554zfxsppi1atKj35R999NHAR3GmT5++cuXKkydP+vn5CYVC41UbNmzo\nUxTFtPHjxz/33HPx8fHbtm2z8FIy/IwCW2DtJgYrM34PjB8//oMPPjDc3L9/v6+vr1WiZs+e\nvWnTph5XrV279oUXXjA/im7a2LFjCSHjxo17zgghZNZ/WSWK88gv62sUrTTuLZGVlfXII49I\npdI333yzra2tH4PhbRTYBhQh64zbSywWG/9GaGlpcXR0tEqURCL54Ycfelz1/fffu7m5mR9F\nN+3+/fuLFi1yd3cvKyszLOzfH5QUo3jLsDl37tyJi4sTCoUSiSQuLq60tLSjo8MGosA22NS/\nOugH49+8v/nNb4wL4/r16yNGjLBWlFar7XGVRqPpa1vQTdPr9bm5uUOHDpXL5Q8ePNBb1l5U\nok6fPt3V1dXvMTy8NJPNuXXr1saNG0eOHEkIEQqFzz777GCPAtuAo0ZZJxAI5syZM2bMmDFj\nxpw8edLNze3IkSOEkLq6uuXLlw8bNuzQoUNWierlndnXa4bRTeN88803UVFRzs7OCoXi8ccf\nt+TfkeVRAoHA29s7Nzd36tSp/R7Gw0jr8bXVaDTV1dWlpaUXLlw4c+bMoI4C24AiZN3Ro0dv\n3rz5z/+qr6/npqERCASenp5nzpx58sknBz6K/0VICFGpVOvWrSssLGxpabHw35GFUQKBIDw8\nvLi4mMrRHxTTKM4vwc8osA14Q8D/UKlU3HVACgoKXnjhhWHDhlklSiAQmJx/beyrr77qaxFS\nTDNx9OjRsrIy86c+eBhR3G/2/fv3c5fxlMvliYmJhtNFrJiWmZmZmJjYv2EMiiiwDShC1vHz\nPGVuTpxe/O1vf7NKGj9fLsMuTlNT0+bNm/Py8hwdHaOjoyMiImbOnGmY/GHg02pqaqh8Wsvb\nKLARA/FFJPAYwXnKfcHPl4tQPfqDYppJVFpaWp9Gwv8osA0oQtYRQsLDw+3s7FavXt3U1MST\nKD4fBsnDl6vHv2gfPHhQUVEhl8sDAgKslWYSZclf3vyMAtuAdwDrCC/PU6a7c0l934uHL1f/\nHviw0/jZXihCMIF3AOsMvwV4dZ4yxb0lumn8fLnefffdPt1/wNL42V4oQjCBdwDrHt4XQpZH\n0boI1sPb9+LJy0X3e1mKafxsLxQhmMA7gHUP7wshKlFULoL1MHbjjPHn5eJYePTHwzsshSft\nhSIEEzh9gnX8PE/ZJOqHH37485//nJ+f39TUJBQKfX19L126ZJW0QfFyWZhMMU0gEMhkMsPN\n+vp645uEkNu3bw/qKLANmIaJde+++y4Po0x4eXnt2rXrzTffNFwEy1ppg+Ll4o+oqCjbjgLb\ngD1CIHfv3v3ss8/EYvHs2bMdHBzUavUPP/zQ0dHBzdj35ZdfDnwU3YtgPexLall9hnre7hEC\nDA4D/Vks8Mzly5cfeeQR7s3g7e392Wefubq69u8dQjGKt4dB9ojiv6P+RRGqX3rRTdPr9Rcv\nXtywYYPhZlxcXEVFhS1FwWCHv/VYFxwc3N7e/vrrrxNCtmzZcvXq1SeffDIuLm7UqFESicTd\n3X3ixIkDH9Ujijte/U7rcVr5jz/+eOHChdzPFs5Q378oPu8RJiUl7dq1y8nJqb29nVsik8ka\nGhqSkpJSU1NtIApsgVVrGKxv2LBhP/74I/fztWvXCCHffvut1aN6RPft2r80fs5QTwiRGTG5\nKZPJ+rSNFNO42RazsrKMj87V6XQHDx6USCSFhYWDPQpsA/YIWWdnZ8dNlkQI0Wg0QqFQq9Xa\n2dlZN4ri3hLdNKVSuXr16vLy8g8//DAoKIhb2L99JopR0dHRvd/h4MGDVkmbNGlSfHz8ihUr\nuq/Kzc3Ny8urrKwc1FFgG1CErKN7rDytqKeeeurmzZvjxo0bPny4YeHZs2dnzZrF/fzZZ59Z\nK40QkpeXt3bt2sTExDfeeGPIkCGWbCnFKB5ycnKqq6szftkN7t27N3r06Hv37g3qKLAR1toV\nBZ4gvDxP+f79+4sWLXJ3dy8rK7M8kG4a5/r16z4+PtOmTbt16xZPouge/UElTSwW379/v8dV\nSqVSJBIN9iiwDShC1vGzCDncdzlyufzBgweWB9JN0+v1nZ2dq1at4iYctnrUpk2bCCFOTk6G\nJR4eHoQQbnJda6X5+PgUFxf3uKq0tPSpp54a7FFgG1CErCOEuBkxuenm5maVKAOKO17U0zhH\njhyJj4+3bhTdoz8opqWmpj7xxBN1dXUmy3/66aenn346OTl5sEeBbbCpLySgH3o8ZMBYdnb2\nwEcZU6lU69atKywsbGlpsfztamFaenr6hg0bLBwD9Si6R39QTOvs7AwICPjmm28WL148bdo0\nV1fXxsbGK1eu5OTkPP744zU1NUOHDh3UUWAjrFzEAOahuONlSRrFfzIUo8Ri8b///e8eV7W2\ntkqlUiumtbW1xcfHi0Qiw+8ckUi0bNmy5ubmPuXwNgpsAPYIWffJJ5/83//9n0Qi6XHtiRMn\n5s+fP/BRFPeW6Kbx86LbTk5OTU1NPe7HtLe3u7q6dnV1WSuNc//+/S+//LKtrc3V1dXHx8fZ\n2bmvCTyPgkENRcg6gUDg7e195MiR8ePH97jW/HcI3SiK70x+thfFKF9f3+3bt4eEhHRfVVZW\ntnbt2hs3blglTaVS9X4HR0fHQR0FtgGzTwBpbGycPHlydnb2r55JPZBRvNXj6fkGlp/p34+o\n2NjYdevWPfPMM56ensbLGxsb169fHxkZaf6Q6KaJxeLe72D+nwL8jALbgD1C1gkEgu+++y4i\nIuLq1at//OMf09LShEKh8do+7cZRjOLtHiF3QOwvaW5uHvgoukd/UEwTCAT29vb+/v7BwcEO\nDg7d/y8kJiYO6iiwDShC1nEloVKp1q5dm52dPX369MOHD3MnjZF+tRetKMP1z3rUpx0vimn8\n/GiUEHL//v2NGzfm5uaq1WpuiUgkio2NTU9P771uH2rasWPHFArFiRMn7O3tQ0NDY2JigoKC\njP88GuxRYBtQhKwz/nWcn5+/evVqqVR68OBBf39/YsFunOVRtPaW6Kbxtgg5dI/+oJWmVCqL\ni4sVCkV5eblUKo2MjIyJiZk5c6ZAILCNKBj0HvJRqcB3Ju+Br776auzYsfb29rt27eq+1lpR\nFqKYxs+ozl9jxTRjzc3NWVlZs2bNsrOze+yxx1599VUbi4JBCkXIuu6/ju/du7dgwQJCSHh4\nuIXtRTHKEhTT+jQ70oBF0f1792H/9dze3p6enu7g4GCrUTDo4KhRMCWVSouKijIyMrgLTvIk\niif6Ok/FwEQRQno/+sO6aRyNRlNWVlZQUFBSUqLRaObPnx8bG2tLUTCIWbuJwcr+9a9//dKq\ns2fPLlq0yCpRFPeWqKe1trb+9a9/PX36tEaj0ev1aWlpEyZMeOaZZ/Lz860VdfTo0cjISLFY\nPHTo0EWLFn3yySdqtbqvg3lIaTqdrqqqavXq1cOHD7e3tw8KCsrNzb13757NRIENQBEC9MGl\nS5dcXV25PyL9/Pzef/99oVD44osvzpkzx87OrqCgwCpRnPv373/44YfBwcFCodDNzW3VqlWV\nlZU6na6vORTTkpKSRo8eTQiZMmXKO++809jY2L/B8DYKbAOOGgXog1mzZul0upSUFK1Wm5KS\ncuXKlaysLO761G+//fbBgwcvXrw48FEm7t69W1RUpFAoqqqqPDw8oqKi0tPT+xdlYRp3xl5A\nQMDo0aO53UqTO/TpxBUeRoFtQBEC9IGTk1Ntbe2YMWMIITdu3PD29v7555+5eQSbmpqeeOKJ\n9vb2gY/qUUdHx759+5KTk7u6uiz/Z96/tAkTJvR+h9ra2kEdBbYBRQjQBwKBQKvV2tnZEUK0\nWu2QIUOM/wX19VxJWlHGTI7+mDdvXmxsLHfUrtXTAPgJRQjQByb91PvNAYsihOj1+rNnzxYU\nFBw+fLilpSUwMJBrLKlUan7IQ0oD4DkUIUAf8LMI5XK5QqGoq6ubMmVKbGxsVFSUu7u7mY99\n2GkA/IciBOgDgUDAXSKAc+TIEZObfSpCilEUj/7AsSTAGhQhQB/w85ANukd/4FgSYA2KEAAA\nmGZn7QEAAABYE4oQAACYhiIEAACmoQgBAIBpKEIAAGAaihAAAJiGIgQAAKahCAEAgGkoQgAA\nYBqKEAAAmIYiBAAApqEIAQCAaShCAABgGooQAACYhiIEAACmoQgBAIBpKEIAAGAaihAAAJiG\nIgQAAKahCAEAgGkoQgAAYBqKEAAAmIYiBAAApqEIAQCAaShCAABgGooQAACYhiIEAACmoQgB\nAIBpKEIAAGAaihAAAJiGIgQAAKahCAEAgGkoQgAAYBqKEAAAmIYiBAAApqEIAQCAaShCAABg\nGooQAACYhiIEAACmoQgBAIBpKEIAAGAaihAAAJiGIgQAAKahCAEAgGkoQgAAYBqKEAAAmIYi\nBAAApqEIAQCAaShCAABgGooQAACYhiIEAACmoQgBAIBpKEIAAGAaihAAAJiGIgQAAKahCAEA\ngGkoQgAAYBqKEAAAmIYiBAAApqEIAQCAaShCAABgGooQAACYhiIEAACmoQgBAIBpKEIAAGAa\nihAAAJiGIgQAAKahCAEAgGkoQgAAYBqKEAAAmIYiBAAApqEIAQCAaShCAABgGooQAACYhiIE\nAACmoQgBAIBpKEIAAGAaihAAAJiGIgQAAKahCAEAgGkoQgAAYBqKEAAAmIYiBAAApqEIAQCA\naShCAABgGooQAACYhiIEAACmoQgBAIBpKEIAAGAaihAAAJiGIgQAAKahCAEAgGkoQgAAYNr/\nBysjmi53VHUkAAAAAElFTkSuQmCC",
      "text/plain": [
       "Plot with title “Color Key\n",
       "and Histogram”"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "options(repr.plot.width=5, repr.plot.height=5)\n",
    "# the next line does not actually work, because the BAM files are not included in the package\n",
    "# ta <- dba.count(ta, minOverlap=3)\n",
    "# instead we load the counts:\n",
    "data(tamoxifen_counts)\n",
    "plot(tamoxifen)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We can perform a test by specifying to contrast over the levels of condition. This will call DESeq software in order to normalize samples for sequencing depth and perform essentially the same analysis as a differential expression analysis for RNA-Seq counts:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "DESeq2"
      ],
      "text/latex": [
       "DESeq2"
      ],
      "text/markdown": [
       "DESeq2"
      ],
      "text/plain": [
       "[1] DESeq2\n",
       "Levels: DESeq2"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "tamoxifen$config$AnalysisMethod"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The plot produced then looks at correlation only for those peaks which showed evidence of differential binding."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "converting counts to integer mode\n",
      "gene-wise dispersion estimates\n",
      "mean-dispersion relationship\n",
      "final dispersion estimates\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "11 Samples, 2845 sites in matrix:\n",
       "       ID Tissue Factor  Condition  Treatment Replicate Caller Intervals FRiP\n",
       "1  BT4741  BT474     ER  Resistant Full-Media         1 counts      2845 0.16\n",
       "2  BT4742  BT474     ER  Resistant Full-Media         2 counts      2845 0.15\n",
       "3   MCF71   MCF7     ER Responsive Full-Media         1 counts      2845 0.27\n",
       "4   MCF72   MCF7     ER Responsive Full-Media         2 counts      2845 0.17\n",
       "5   MCF73   MCF7     ER Responsive Full-Media         3 counts      2845 0.23\n",
       "6   T47D1   T47D     ER Responsive Full-Media         1 counts      2845 0.10\n",
       "7   T47D2   T47D     ER Responsive Full-Media         2 counts      2845 0.06\n",
       "8  MCF7r1   MCF7     ER  Resistant Full-Media         1 counts      2845 0.20\n",
       "9  MCF7r2   MCF7     ER  Resistant Full-Media         2 counts      2845 0.13\n",
       "10  ZR751   ZR75     ER Responsive Full-Media         1 counts      2845 0.32\n",
       "11  ZR752   ZR75     ER Responsive Full-Media         2 counts      2845 0.22\n",
       "\n",
       "1 Contrast:\n",
       "     Group1 Members1     Group2 Members2 DB.DESeq2\n",
       "1 Resistant        4 Responsive        7       629"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "ta2 <- dba.contrast(tamoxifen, categories=DBA_CONDITION)\n",
    "ta2 <- dba.analyze(ta2)\n",
    "ta2"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "*Note*: We could have included the tissue as a blocking factor, by providing `DBA_TISSUE` to the `block` argument of `dba.contrast`.\n",
    "\n",
    "From the `DiffBind` vignette, we have:\n",
    "\n",
    "> By default, dba.analyze plots a correlation heatmap if it finds any \n",
    "> significantly differentially bound sites, shown in Figure 3. Using only \n",
    "> the differentially bound sites, we now see that the four tamoxifen \n",
    "> resistant samples (representing two cell lines) cluster together, \n",
    "> although the tamoxifen-responsive MCF7 replicates cluster closer to them \n",
    "> than to the other tamoxifen responsive samples.\"\n",
    "\n",
    "Finally, we can generate the results table, which is attached as metadata columns to the peaks as genomic ranges. By specifying `bCounts = TRUE`, we also obtain the normalized counts for each sample."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "GRanges object with 629 ranges and 6 metadata columns:\n",
       "       seqnames               ranges strand |      Conc Conc_Resistant\n",
       "          <Rle>            <IRanges>  <Rle> | <numeric>      <numeric>\n",
       "  2452    chr18 [64490686, 64491186]      * |      6.36           1.39\n",
       "  1291    chr18 [34597713, 34598213]      * |      5.33           0.22\n",
       "   976    chr18 [26860997, 26861497]      * |       7.3           3.13\n",
       "  2338    chr18 [60892900, 60893400]      * |      7.13           1.84\n",
       "  2077    chr18 [55569087, 55569587]      * |      5.52           1.89\n",
       "   ...      ...                  ...    ... .       ...            ...\n",
       "   551    chr18 [14465945, 14466445]      * |      6.02           4.38\n",
       "  2659    chr18 [71909888, 71910388]      * |      5.58           3.73\n",
       "  2541    chr18 [68007206, 68007706]      * |      3.61           2.41\n",
       "  1967    chr18 [52609747, 52610247]      * |      3.87           2.39\n",
       "  2383    chr18 [61927095, 61927595]      * |      1.72          -0.22\n",
       "       Conc_Responsive      Fold   p-value       FDR\n",
       "             <numeric> <numeric> <numeric> <numeric>\n",
       "  2452               7     -5.61  3.57e-10  1.02e-06\n",
       "  1291            5.97     -5.75   1.1e-09  1.57e-06\n",
       "   976            7.92     -4.79   1.1e-08  1.05e-05\n",
       "  2338            7.77     -5.93  1.68e-08  1.17e-05\n",
       "  2077            6.13     -4.23  2.36e-08  1.17e-05\n",
       "   ...             ...       ...       ...       ...\n",
       "   551            6.49     -2.11    0.0108     0.049\n",
       "  2659            6.07     -2.34    0.0108    0.0491\n",
       "  2541            4.01      -1.6    0.0109    0.0495\n",
       "  1967            4.32     -1.94    0.0109    0.0495\n",
       "  2383            2.22     -2.44     0.011    0.0498\n",
       "  -------\n",
       "  seqinfo: 1 sequence from an unspecified genome; no seqlengths"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "tadb <- dba.report(ta2)\n",
    "tadb\n",
    "counts <- dba.report(ta2, bCounts=TRUE)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##  Annotating and plotting peaks\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "First, do any differentially bound peaks overlap with genes? Yes, 299 of the 629 peaks do:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Loading required package: AnnotationDbi\n",
      "Loading required package: OrganismDbi\n",
      "Loading required package: GenomicFeatures\n",
      "Loading required package: GO.db\n",
      "Loading required package: org.Hs.eg.db\n",
      "\n",
      "Loading required package: TxDb.Hsapiens.UCSC.hg19.knownGene\n",
      "'select()' returned 1:1 mapping between keys and columns\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "   Mode   FALSE    TRUE \n",
       "logical     330     299 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "library(Homo.sapiens)\n",
    "gn <- genes(Homo.sapiens, columns=\"SYMBOL\")\n",
    "summary(counts %over% gn)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "How many genes does each peak overlap with?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "\n",
       "  0   1   2 \n",
       "330 289  10 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "table(countOverlaps(counts, gn))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Let's look at those peaks that overlap with two genes:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<ol class=list-inline>\n",
       "\t<li>107.743</li>\n",
       "\t<li>325.387</li>\n",
       "\t<li>18.89</li>\n",
       "\t<li>69.018</li>\n",
       "\t<li>319.503</li>\n",
       "\t<li>46.973</li>\n",
       "\t<li>156.669</li>\n",
       "\t<li>269.708</li>\n",
       "\t<li>959.237</li>\n",
       "</ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 107.743\n",
       "\\item 325.387\n",
       "\\item 18.89\n",
       "\\item 69.018\n",
       "\\item 319.503\n",
       "\\item 46.973\n",
       "\\item 156.669\n",
       "\\item 269.708\n",
       "\\item 959.237\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 107.743\n",
       "2. 325.387\n",
       "3. 18.89\n",
       "4. 69.018\n",
       "5. 319.503\n",
       "6. 46.973\n",
       "7. 156.669\n",
       "8. 269.708\n",
       "9. 959.237\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "[1] 107.743 325.387  18.890  69.018 319.503  46.973 156.669 269.708 959.237"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "GRanges object with 9 ranges and 1 metadata column:\n",
       "            seqnames               ranges strand |          SYMBOL\n",
       "               <Rle>            <IRanges>  <Rle> | <CharacterList>\n",
       "       9229    chr18 [ 3496030,  4455266]      - |          DLGAP1\n",
       "     201477    chr18 [ 3878180,  3897069]      + |      DLGAP1-AS3\n",
       "     147429    chr18 [24445272, 24770658]      + |        AQP4-AS1\n",
       "      83539    chr18 [24495595, 24765302]      - |           CHST9\n",
       "  100505549    chr18 [55297534, 55405276]      + |    LOC100505549\n",
       "       5205    chr18 [55313659, 55470327]      - |          ATP8B1\n",
       "     221241    chr18 [61747243, 61816260]      - |       LINC00305\n",
       "     284294    chr18 [61771325, 62090827]      + |       LINC01924\n",
       "     400654    chr18 [61880318, 61927290]      - |       LINC01538\n",
       "  -------\n",
       "  seqinfo: 93 sequences (1 circular) from hg19 genome"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "count2 <- counts[countOverlaps(counts, gn) == 2]\n",
    "gn2 <- gn[gn %over% count2]\n",
    "width(gn2) / 1e3  #width in kb\n",
    "gn2 <- gn2[order(ranges(gn2))]\n",
    "gn2"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Plot peaks in the UCSC genome browser\n",
    "Although the `DiffBind` package doesn't specify the genome, let's assume it is `hg19`, add a track to the `hg19` genome in the UCSC genome browser, then start the browser centered on the peak with greatest fold-change:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "genome(counts) <- \"hg19\"\n",
    "library(rtracklayer)\n",
    "session <- browserSession(\"UCSC\")\n",
    "genome(session) <- \"hg19\"\n",
    "track(session, \"counts\") <- counts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "UCSCView of :33031597-33041570 \n",
       "trackNames(10): 'counts' 'Base Position' ... 'Common SNPs(150)' 'RepeatMasker'"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "rangemaxFC <- counts[which.max(abs(counts$Fold))]\n",
    "browserView(session, range=rangemaxFC * 0.75)  #0.75 zoom-factor"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Finally, let's plot the peak that overlapped with *AQP4-AS1* on the sense strand and *CHST9* on the antisense strand:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "GRanges object with 2 ranges and 1 metadata column:\n",
       "         seqnames               ranges strand |          SYMBOL\n",
       "            <Rle>            <IRanges>  <Rle> | <CharacterList>\n",
       "  147429    chr18 [24445272, 24770658]      + |        AQP4-AS1\n",
       "   83539    chr18 [24495595, 24765302]      - |           CHST9\n",
       "  -------\n",
       "  seqinfo: 93 sequences (1 circular) from hg19 genome"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "gn2[3:4]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "GRanges object with 1 range and 0 metadata columns:\n",
       "      seqnames               ranges strand\n",
       "         <Rle>            <IRanges>  <Rle>\n",
       "  [1]    chr18 [24445272, 24770658]      *\n",
       "  -------\n",
       "  seqinfo: 93 sequences (1 circular) from hg19 genome"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "(plotrange <- reduce(gn2[3:4], ignore.strand=TRUE))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "UCSCView of :69433024-69433691 \n",
       "trackNames(10): 'counts' 'Base Position' ... 'Common SNPs(150)' 'RepeatMasker'"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "browserView(session, range=plotrange * 0.75)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Footnotes <a name=\"foot\"></a>\n",
    "\n",
    "### Model-based Analysis for ChIP-Seq (MACS)\n",
    "\n",
    "Zhang Y, Liu T, Meyer CA, Eeckhoute J, Johnson DS, Bernstein BE, Nusbaum C, Myers RM, Brown M, Li W, Liu XS. \"Model-based Analysis of ChIP-Seq (MACS)\". Genome Biol. 2008.\n",
    "<http://www.ncbi.nlm.nih.gov/pmc/articles/PMC2592715/>\n",
    "\n",
    "Software: \n",
    "\n",
    "<http://liulab.dfci.harvard.edu/MACS/>\n",
    "\n",
    "### Motif finding\n",
    "\n",
    "Wikipedia's article on DNA sequence motifs: <http://en.wikipedia.org/wiki/Sequence_motif>\n",
    "\n",
    "A non-comprehensive list of software for motif finding:\n",
    "\n",
    "- [MEME/DREME](http://meme.nbcr.net/meme/)\n",
    "- [RSAT peak-motifs](http://rsat.ulb.ac.be/peak-motifs_form.cgi)\n",
    "- [motifRG (Bioconductor)](http://www.bioconductor.org/packages/release/bioc/html/motifRG.html)\n",
    "- [rGADEM (Bioconductor)](http://www.bioconductor.org/packages/release/bioc/html/rGADEM.html)\n",
    "\n",
    "A survey of motif finding algorithms: <http://www.biomedcentral.com/1471-2105/8/S7/S21/>\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "3.4.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
